---
title: "MB - TB analysis"
author: "Alice Risely"
date: "25/04/2022"
output:
  html_document:
    toc: yes
    toc_depth: 3
    number_sections: true
    toc_float:
      collapsed: false
    theme: journal
    df_print: paged
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

## Load packages

```{r load packages}

library(phyloseq)
library(ggord)
library(ggplot2)
library(viridis)
library(mgcViz)
library(dichromat)
library(ggpattern)
library(expss)
library(tidyverse)
library(GGally)
library(vegan)
library(ggpubr)
library(pheatmap)
library(lubridate)
library(ggsci)
library(RColorBrewer)
library(microbiome)
library(vegan)
library(lme4)
library(performance)
library(ggrepel)

library(glmmTMB)
library(speedyseq)
library(ggeffects)
library(MuMIn)
library(effects)
library(broom)

library("coxme")
library(survival)

library(SpiecEasi)
library(NetCoMi)

memory.limit(1000000)

```

# 1. Data import and cleaning

```{r}

meerkat_final<-readRDS("C:\\Users\\risel\\Dropbox\\Sommer postdoc\\Meerkat project\\PROJECTS\\MeerkatSpikeInData\\Analysis updated\\DATA 2021\\meerkat_unrarefied.rds")

sample_data(meerkat_final)<-sample_data(meerkat_final)[,1:56]

sample_data(meerkat_final)$Observed_unnormalised<-as.numeric(estimate_richness(meerkat_final)$Observed)

asv_table<-data.frame(meerkat_final@otu_table@.Data)
asv_table[1:5,1:5]
names(asv_table)<-sample_data(meerkat_final)$feature.id
meerkat_normalized<-sweep(asv_table, MARGIN=2, sample_data(meerkat_final)$scaling_factor, `*`)
meerkat_normalized<-round(meerkat_normalized, digits = 0)
meerkat_normalized<-otu_table(meerkat_normalized, taxa_are_rows = TRUE)
meerkat_scaled<-meerkat_final
otu_table(meerkat_scaled)<-meerkat_normalized

# remove mhc variables
names(sample_data(meerkat_scaled))

### add total abundance of reads to dataframe
sample_data(meerkat_scaled)$TotalAbundance<-sample_sums(meerkat_scaled)

meerkat_scaled



#Filter samples

#sample_data(meerkat_scaled)$MHC_data<-is.na(sample_data(meerkat_scaled)$AA_pdist_mean)
#table(sample_data(meerkat_scaled)$MHC_data)

#meerkat_filtered<-subset_samples(meerkat_scaled, MHC_data == FALSE)
meerkat_filtered<-subset_samples(meerkat_scaled, Seq_depth_nospikein >3000)
meerkat_filtered<-subset_samples(meerkat_filtered, sample_sums(meerkat_filtered)>1000)


### change some variables 

sample_data(meerkat_filtered)$TimeCat<-ifelse(sample_data(meerkat_filtered)$HoursAfterSunrise < 7, "MORNING", "AFTERNOON")
table(sample_data(meerkat_filtered)$TimeCat)
meerkat_filtered

head(sample_data(meerkat_filtered))



# generate categorical variable for wet and dry seasons (May-Sep = dry, Oct-April = Wet)

sample_data(meerkat_filtered)$month <-lubridate::month(sample_data(meerkat_filtered)$SampleDate)

sample_data(meerkat_filtered)$Season <- as.factor(ifelse(sample_data(meerkat_filtered)$month == 5| 
    sample_data(meerkat_filtered)$month == 6 |
    sample_data(meerkat_filtered)$month == 7 |
    sample_data(meerkat_filtered)$month == 8 |
    sample_data(meerkat_filtered)$month == 9  , "DRY", "WET"))


metadata<-data.frame(sample_data(meerkat_filtered))

ggplot(metadata, aes(x = SampleDate, y = reorder(IndividID, SampleDate), group = IndividID))+geom_point(aes(col  = TB_stage))+geom_line()

ggplot(metadata, aes(x = SampleDate, y =  reorder(IndividID, SampleDate), group = IndividID))+geom_point(aes(col  = Storage))+geom_line()#


summary(sample_data(meerkat_filtered)$AgeAtDeath)
hist(sample_data(meerkat_filtered)$AgeAtDeath, breaks = 50)

## TB status

levels(sample_data(meerkat_filtered)$TB_status)
levels(sample_data(meerkat_filtered)$TB_stage)

sample_data(meerkat_filtered)$TB_status<- factor(sample_data(meerkat_filtered)$TB_status, levels = c("unexposed", "exposed_asymptomatic", "exposed_symptomatic"))

levels(sample_data(meerkat_filtered)$TB_status)<-c("Unexposed", "Exposed", "Diseased")


### TB stage

sample_data(meerkat_filtered)$TB_stage<- factor(sample_data(meerkat_filtered)$TB_stage, levels = c("TB_uninfected", "TB_exposed", "TB_symptomatic"))

levels(sample_data(meerkat_filtered)$TB_stage)<-c("Unexposed", "Exposed", "Diseased")



########## alpha diversity

sample_data(meerkat_filtered)$Observed<-as.numeric(estimate_richness(meerkat_filtered)$Observed)
sample_data(meerkat_filtered)$Shannon<-as.numeric(estimate_richness(meerkat_filtered)$Shannon)

ggplot(sample_data(meerkat_filtered), aes(x = Observed_unnormalised, y = Observed))+geom_point()


# group level TB status ########
# group level TB status ########
# group level TB status ########
# group level TB status ########

TB_groups <- read.csv("C:/Users/risel/Dropbox/Sommer postdoc/Meerkat project/PROJECTS/MEERKAT DATA/TB data/TB_symptom_dates_group_modified.csv")


# group abbreviations

group_history<-read.csv("C:/Users/risel/Dropbox/Sommer postdoc/Meerkat project/PROJECTS/MEERKAT DATA/Meerkatdatabase/tblGroups.csv")[,c(2:9)]


head(group_history)
head(TB_groups)

group_abbreviations<-group_history[,c(2:3)]

TB_groups<-merge(TB_groups, group_abbreviations, by = "GroupAbbrev", all.x = T)


TB_groups$TB<-as.Date(TB_groups$TB, format = "%d/%m/%Y")
TB_groups$TB_suspected<-as.Date(TB_groups$TB_suspected, format = "%d/%m/%Y")
TB_groups$TB_Josi<-as.Date(TB_groups$TB_Josi, format = "%d/%m/%Y")
names(TB_groups)[2]<-"TB_group"
names(TB_groups)[3]<-"TB_suspected_group"



metadata<-data.frame(sample_data(meerkat_filtered))
head(metadata)
metadata<-merge(metadata, TB_groups[,c(1:3,5)], by.x = "GroupAtSampling", by.y = "GroupName")

head(metadata)

row.names(metadata)<-metadata$feature.id

sample_data(meerkat_filtered)<-metadata

sample_data(meerkat_filtered)$TB_present_group<- ifelse(sample_data(meerkat_filtered)$SampleDate < sample_data(meerkat_filtered)$TB_group, FALSE, TRUE)

#sample_data(meerkat_filtered)$TB_present_group<- ifelse(sample_data(meerkat_filtered)$SampleDate < sample_data(meerkat_filtered)$TB_suspected_group, FALSE, TRUE)

table(sample_data(meerkat_filtered)$TB_present_group)


# add missing death dates for 3 meerkats 
# add missing death dates for 3 meerkats 
# add missing death dates for 3 meerkats 
# add missing death dates for 3 meerkats 

#210: 3/11/2007
#211: 4/08/2008
#3111: 20/02/2019

sample_data(meerkat_filtered)$DeathDate<-data.table::fifelse(sample_data(meerkat_filtered)$IndividID == 210, as.Date("2007-11-03"), as.Date(sample_data(meerkat_filtered)$DeathDate))

sample_data(meerkat_filtered)$DeathDate<-data.table::fifelse(sample_data(meerkat_filtered)$IndividID == 211, as.Date("2008-08-04"), as.Date(sample_data(meerkat_filtered)$DeathDate))

sample_data(meerkat_filtered)$DeathDate<-data.table::fifelse(sample_data(meerkat_filtered)$IndividID == 3111, as.Date("2019-02-20"), as.Date(sample_data(meerkat_filtered)$DeathDate))

```


# 3. Fig. S1. Sampling design

```{r}


p1<-ggplot(data=sample_data(meerkat_filtered),aes(x=SampleDate,y=reorder(IndividID, SampleDate),group=IndividID))+
  geom_line(size=0.1,alpha=0.5, col = "darkgrey")+
  geom_point(size=2, aes(fill = TB_stage), pch = 21)+
   geom_point(data = subset(sample_data(meerkat_filtered), TB_stage == "Diseased"),size=2,  fill = "brown2", pch = 21)+
  scale_fill_manual(values = c("forestgreen", "skyblue", "brown2"))+
  theme_bw(base_size = 14)+
  theme(strip.background = element_blank(), strip.text = element_blank())+
  theme(panel.grid.minor=element_blank(),panel.grid.major=element_blank(),
        panel.background=element_blank())+
  theme(axis.text=element_blank(),axis.title.x = element_blank(),
        axis.ticks.y=element_blank())+
  scale_x_date(date_breaks = "5 year")+
  #theme(axis.text.x = element_text(angle = 45, vjust = 0.9, hjust=1))+
  ylab("Meerkat ID")+
  #scale_y_discrete(expand=expansion(add=c(0.2,1.5)))+

   geom_vline(xintercept = as.Date("2000-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2005-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2010-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2015-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2020-01-01"), linetype = "dashed", col = "grey")+
  coord_cartesian(xlim = c(as.Date("1997-01-01"), as.Date("2020-01-01")))+
  theme(plot.margin=unit(c(0.2,0.3,0,0.2),"cm"))+
   guides(fill = guide_legend(override.aes = list(size = 3)))



TB_MHC_data <- read.csv("C:/Users/risel/Dropbox/Sommer postdoc/Meerkat project/PROJECTS/MEERKAT DATA/TB data/TB_MHC_data.csv")

head(TB_MHC_data)
dim(TB_MHC_data)
names(TB_MHC_data)
ind_data<-TB_MHC_data[,1:21]
hist(ind_data$Final_age)
table(ind_data$Birth_Group)
ind_data$ReachedAdulthood<-ind_data$Final_age>365
table(ind_data$ReachedAdulthood)

ind_data2<-subset(ind_data, ReachedAdulthood==T)
head(ind_data2)

group_freq<-data.frame(table(ind_data2$Birth_Group))
group_freq<-group_freq %>% arrange(-Freq)
head(group_freq, 20)

top_groups<-group_freq$Var1[1:20]

ind_data2$TopGroup<-ind_data2$Birth_Group %in% top_groups

ind_data3<-subset(ind_data2, TopGroup == T)
names(ind_data3)

dim(ind_data3)
ind_data4<-ind_data3[,c(1,11,7)]

ind_data5 <- gather(ind_data4, Type, Date, BirthDate:DeathDate, factor_key=TRUE)

ind_data5$Date<-as.Date(ind_data5$Date, format = "%d/%m/%Y")

ind_data5$IndividID<-factor(ind_data5$IndividID)

ggplot(ind_data5, aes(y = IndividID, x = Date, group = IndividID))+geom_point()+geom_line()

ind_data6<-merge(ind_data5, ind_data3, by = "IndividID")


ggplot(ind_data6, aes(y = reorder(IndividID, Date), x = Date, group = IndividID))+geom_line()+facet_wrap(~Birth_Group, scales = "free_y")

ind_data6$Microbiome<- ind_data6$IndividID %in% sample_data(meerkat_filtered)$IndividID

ggplot(ind_data6, aes(y = reorder(IndividID, Date), x = Date, group = IndividID, col = Microbiome))+geom_line()+facet_wrap(~Birth_Group, scales = "free_y")


ind_data4$Microbiome<- ind_data4$IndividID %in% sample_data(meerkat_filtered)$IndividID

table(ind_data4$Microbiome)

ind_data6<-subset(ind_data6, !is.na(Date))
str(ind_data6)

ind_data6$IndividID<-reorder(ind_data6$IndividID, ind_data6$Date, min)

p2<-ggplot(ind_data6, aes(y = reorder(IndividID, Date), x = Date, group = IndividID, col = Microbiome))+
  geom_line(alpha = 0.4, size = 0.1)+
 # facet_wrap(~reorder(Birth_Group, Date, FUN =  min), scales = "free_y", ncol = 4)+
  theme_bw(base_size = 14)+
  theme(strip.background = element_blank(), strip.text = element_blank())+
  scale_color_manual(values = c("darkgrey", "red"))+
  geom_line(data = subset(ind_data6, Microbiome == T), col = "red", alpha = 0.6)+
  theme(axis.text.y = element_blank(), 
        axis.ticks.y = element_blank())+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())+
   ylab("Meerkat ID")+
  geom_vline(xintercept = as.Date("2000-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2005-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2010-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2015-01-01"), linetype = "dashed", col = "grey")+
   geom_vline(xintercept = as.Date("2020-01-01"), linetype = "dashed", col = "grey")+
  coord_cartesian(xlim = c(as.Date("1997-01-01"), as.Date("2020-01-01")))+
  guides(colour = guide_legend(override.aes = list(alpha = 1, size = 2)))


ind_data6$TB_status<-factor(ind_data6$TB_status, levels = c("unexposed", "exposed_asymptomatic", "exposed_symptomatic"))
levels(ind_data6$TB_status)<-c("Unexposed", "Exposed", "Diseased")

p3<-ggplot(ind_data6, aes(y = reorder(IndividID,Date), x = Date, group = IndividID, col = TB_status))+
  geom_line(alpha = 0.4, size = 0.1)+
  geom_line(data = subset(ind_data6, TB_status == "Unexposed"), alpha = 0.4, size = 0.1, col = "skyblue")+
#  facet_wrap(~reorder(Birth_Group, Date, FUN =  min), scales = "free_y", ncol = 4)+
  theme_bw()+
  theme(strip.background = element_blank(), strip.text = element_blank())+
  scale_color_manual(values = c("skyblue", "goldenrod1", "red"))+
  #theme_minimal()+
  theme(axis.text.y = element_blank(), axis.ticks.y = element_blank())+
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank())+
  ylab("Meerkat ID")+
  geom_vline(xintercept = as.Date("2000-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2005-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2010-01-01"), linetype = "dashed", col = "grey")+
  geom_vline(xintercept = as.Date("2015-01-01"), linetype = "dashed", col = "grey")+
   geom_vline(xintercept = as.Date("2020-01-01"), linetype = "dashed", col = "grey")+
  coord_cartesian(xlim = c(as.Date("1997-01-01"), as.Date("2020-01-01")))+
  guides(colour = guide_legend(override.aes = list(alpha = 1, size = 2)))



#figS1bottom<-ggarrange(p2,p3, align = "v", ncol = 1, labels= c("b)", "c)"))

#ggarrange(p1, figS1bottom, ncol = 1, heights = c(1,1.8), labels = c("a)",NA))


ggarrange(p1,p2, ncol = 1, heights = c(1,1), labels = c("a)","b)"))

```

# Fig. 1a-d: Yearly trends

-   Weather data

```{r}

daily_temp<-read.csv("C:/Users/risel/Dropbox/Sommer postdoc/Meerkat project/PROJECTS/MEERKAT DATA/Weather data/SA Weather service data/formatted_daily_temp_VanZylsrus_1997_2019.csv")[-1]

daily_rainfall<-read.csv("C:/Users/risel/Dropbox/Sommer postdoc/Meerkat project/PROJECTS/MEERKAT DATA/Weather data/SA Weather service data/formatted_daily_rainfall_VanZylsrus_1997_2019.csv")[-1]

daily_rainfall$Date<-as.Date(daily_rainfall$Date, "%Y-%m-%d")
daily_temp$Date<-as.Date(daily_temp$Date, "%Y-%m-%d")

head(daily_rainfall)
head(daily_temp)

daily_temp<-na.omit(daily_temp)


##### make HydroYear

daily_rainfall$HydroDate<-daily_rainfall$Date - 91
daily_temp$HydroDate<-daily_temp$Date - 91

daily_rainfall$HydroYear<-year(daily_rainfall$HydroDate)
daily_temp$HydroYear<-year(daily_temp$HydroDate)

daily_rainfall$Year<-lubridate::year(daily_rainfall$Date)
daily_temp$Year<-lubridate::year(daily_temp$Date)

ggplot(daily_rainfall, aes(x = Date, y = Rainfall.mm.))+geom_col()+
  scale_y_sqrt()


daily_rainfall$Month<-month(daily_rainfall$Date)


ggplot(daily_rainfall, aes(x = Date, y = Rainfall.mm.))+geom_col()+
  scale_y_sqrt()+
  geom_vline(xintercept = as.Date("2000-01-01"), linetype = "dashed")+
    geom_vline(xintercept = as.Date("2001-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2002-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2003-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2004-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2005-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2006-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2007-01-01"), linetype = "dashed")+
    geom_vline(xintercept = as.Date("2008-01-01"), linetype = "dashed")+
   geom_vline(xintercept = as.Date("2009-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2010-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2011-01-01"), linetype = "dashed")+
    geom_vline(xintercept = as.Date("2012-01-01"), linetype = "dashed")+
   geom_vline(xintercept = as.Date("2013-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2014-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2015-01-01"), linetype = "dashed")+
    geom_vline(xintercept = as.Date("2016-01-01"), linetype = "dashed")+
   geom_vline(xintercept = as.Date("2017-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2018-01-01"), linetype = "dashed")+  geom_vline(xintercept = as.Date("2019-01-01"), linetype = "dashed")+
    geom_vline(xintercept = as.Date("2020-01-01"), linetype = "dashed")+
  coord_cartesian(ylim = c(0,50))

monthly_rainfall<-daily_rainfall %>% group_by(Month, Year) %>% summarize(sum_rainfall = sum(Rainfall.mm.))

monthly_rainfall$Month<-as.factor(monthly_rainfall$Month)


ggplot(monthly_rainfall, aes(x = Month, y = sum_rainfall))+geom_col()+
 facet_grid(~Year)+scale_y_sqrt()

```

```{r}

# rainfall


yearly_rainfall_df<-daily_rainfall %>% group_by(Year) %>% summarize(Measurement = mean(Rainfall.mm.), se = sd(Rainfall.mm.)/sqrt(n()))

head(yearly_rainfall_df)

yearly_rainfall_df<-subset(yearly_rainfall_df, Year != 2020)
yearly_rainfall_df<-subset(yearly_rainfall_df, Year != 1996)


## temp ######
## temp ######
## temp ######
## temp ######

yearly_temp_df<-daily_temp %>% group_by(Year, Measure) %>% summarize(Measurement = mean(Temperature),  se = sd(Temperature)/sqrt(n()))

yearly_temp_df<-subset(yearly_temp_df, Year != 2020)
yearly_temp_df<-subset(yearly_temp_df, Year != 1996)
yearly_temp_df<-subset(yearly_temp_df, Measure == "mx")


##### merge

yearly_rainfall_df$Climate <- "Rainfall"
yearly_temp_df$Climate <- "Max temp"

yearly_temp_df<-yearly_temp_df[,-2]

climate_df_Year<-rbind(yearly_rainfall_df, yearly_temp_df)

ggplot(climate_df_Year, aes(x = Year, y = Measurement, col = Climate))+
  geom_point()+
  geom_path()+
  geom_errorbar(aes(min =Measurement-se, max = Measurement + se), width = 0.2)+
  facet_wrap(~Climate, scales = "free_y", ncol = 2)+
  theme_bw()


ggplot(climate_df_Year, aes(x = Year, y = Measurement, col = Climate))+
  geom_point()+
  geom_path()+
  geom_errorbar(aes(min =Measurement-se, max = Measurement + se), width = 0.2)+
  facet_wrap(~Climate, scales = "free_y", ncol = 2)+
  theme_bw()

```

### TB by year

```{r}
#TB_by_year <- read.csv("C:/Users/risel/Dropbox/Sommer postdoc/Meerkat project/PROJECTS/MEERKAT DATA/TB data/TB_by_year_2021_11_12.csv")
TB_by_year <- read.csv("C:/Users/risel/Dropbox/Sommer postdoc/Meerkat project/PROJECTS/MEERKAT DATA/TB data/TB_by_year_ID_and_groups.csv")

head(TB_by_year)

TB_summary_year<-TB_by_year %>% group_by(year, current_TB_status_2) %>% summarize(freq = n())

head(TB_summary_year)
TB_summary_year2<-TB_by_year %>% group_by(year) %>% summarize(freq = n())
head(TB_summary_year2)

TB_summary_year <-subset(TB_summary_year, year > 1996)
TB_summary_year <-subset(TB_summary_year, year <2020)

head(TB_summary_year)

TB_summary_year$Total<- vlookup(TB_summary_year$year, TB_summary_year2, lookup_column = "year", result_column = "freq")


TB_summary_year$Proportion<- TB_summary_year$freq/ TB_summary_year$Total

head(TB_summary_year)

TB_summary_year$se <- sqrt(TB_summary_year$Proportion∗(1−TB_summary_year$Proportion)/TB_summary_year$Total)
names(TB_summary_year)[2]<-"TB_status"


TB_summary_year$TB_status<-factor(TB_summary_year$TB_status, levels = c("unexposed", "exposed", "TB_signs"))

levels(TB_summary_year$TB_status)<-c("TB unexposed", "TB exposed", "TB diseased")

names(TB_summary_year)[1]<-"Year"
str(TB_summary_year)
TB_summary_year<-data.frame(TB_summary_year)



```


## Fig. 1a-d build

```{r}
#str(TB_summary_year)


plot_TB2<-ggplot(subset(TB_summary_year, TB_status != "TB unexposed"), aes(x = Year, y = Proportion, group = TB_status, fill = TB_status, col =TB_status))+
  theme_bw(base_size = 14)+
  geom_path(size = 1)+
   geom_errorbar(aes(min = Proportion -se, max = Proportion +se), width = 0, size = 1)+
  geom_point(size = 3, pch = 21, col = "black")+
  scale_fill_manual(values = c("goldenrod1", "red"))+
  scale_color_manual(values = c("goldenrod1", "red"))+
  ylab("Proportion population")+
  labs(fill = "TB status", col = "TB status")+
    xlab("Year")+
  theme(legend.position = c(0.25, 0.90),
       # legend.justification = c("center", "top"), 
        legend.title = element_blank(), 
         legend.key.height = unit(0.5, 'cm'),
        legend.key.width = unit(0.5, 'cm'),
        legend.text =element_text(size=14))+
     scale_x_continuous(breaks =  c(1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,  2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015,2016, 2017, 2018, 2019), expand = c(0,0))+
theme(axis.text.x = element_text(angle = 90, vjust=0.5))+
  theme( legend.background = element_blank(),
        legend.box.background =element_blank())
          
          #element_rect(colour = "grey", fill = "white"))




yearly_temp_df<-data.frame(yearly_temp_df)
yearly_temp_df$Year<-as.integer(yearly_temp_df$Year)



# temp 

plot_temp2<- ggplot(yearly_temp_df, aes(x = Year, y = Measurement))+
  theme_bw(base_size = 14)+
   geom_smooth(method = "lm", col = "grey")+
  xlab("Year")+
  geom_path(size = 1, col = "grey")+
  geom_errorbar(aes(min = Measurement - se, max = Measurement +se), width = 0)+
   geom_point(size = 2, pch = 21, col = "black", fill = "darkgrey")+
  ylab("Mean max. temp.")+
  
   scale_x_continuous(breaks =  c(1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,  2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015,2016, 2017, 2018, 2019), expand = c(0,0))+
theme(axis.text.x = element_text(angle = 90, vjust=0.5))
# rainfall

plot_rainfall<- ggplot(yearly_rainfall_df, aes(x = Year, y = Measurement))+
  theme_bw(base_size = 14)+
   geom_smooth(method = "lm", col = "grey", alpha = 0.2)+
  xlab("Year")+
  geom_path(size = 1, col = "grey")+
  geom_errorbar(aes(min = Measurement - se, max = Measurement +se), width = 0)+
   geom_point(size = 2, pch = 21, col = "black", fill = "darkgrey")+
  ylab("Mean rainfall")+
     scale_x_continuous(breaks =  c(1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005,  2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015,2016, 2017, 2018, 2019), expand = c(0,0))+
theme(axis.text.x = element_text(angle = 90, vjust=0.5))+
  coord_cartesian(ylim = c(0,1.5))




# Figure 1
ggarrange(plot_temp2, plot_rainfall, plot_TB2, nrow = 1, ncol = 3, labels = c("a)","b)", "c)"))

ggsave("Figures/Fig1.pdf")


```

# 5. Fig. 2e-i: Sample microbiome and metadata

```{r}

meerkat_class<-tax_glom(meerkat_filtered, taxrank = "Class")

#meerkat_class<-subset_samples(meerkat_class, TimeCat == "MORNING")

sample_data(meerkat_class)<-sample_data(meerkat_class)[,c("feature.id")]
meerkat_class<-microbiome::transform(meerkat_class, "compositional") 
class_per_sample<-psmelt(meerkat_class)

head(class_per_sample)

top_class<-c("Bacteroidia","Fusobacteriia" ,  "Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli")


class_per_sample$Class_plot<-as.character(ifelse(class_per_sample$Class %in% top_class, class_per_sample$Class, "Other"))

class_per_sample %>% group_by(Class) %>% summarize(mean = mean(Abundance)) %>% arrange(-mean)



class_per_sample$Class_plot<-factor(class_per_sample$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))

head(class_per_sample)

metadata<-data.frame(sample_data(meerkat_filtered))
names(metadata)

metadata<-metadata[,c("feature.id", "SampleDate", "Year", "Storage", "HoursAfterSunrise", "TotalAbundance", "TB_stage", "Season", "Temp_max",  "sum_rainfall_2weeks", "sum_rainfall_2weeks", "Condition_weight", "GroupAtSampling")]

class_per_sample<- merge(class_per_sample, metadata, by = "feature.id")

## microbiome plot


pal<-brewer.pal(10,"Paired")
pal<-c(pal, "lightgrey")
scales::show_col(pal)
pal<-pal[c(2,1,11,4,3,6,9,10)]
pal<-rev(pal)




plot_per_sample<-ggplot(class_per_sample, aes(y = Abundance, x = reorder(Sample, SampleDate), fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
  scale_fill_manual(values = rev(pal))+
  ylab("Relative abundance")+
  theme(axis.text.x = element_blank(), axis.ticks = element_blank(), axis.title.x = element_blank())+
   scale_y_continuous(expand = c(0,0)) +
  labs(fill = "Bacterial class")+
   theme(plot.margin=unit(c(0.2,0.2,0.2,1),"cm"))+
  theme(legend.justification = c("left", "center"))

ggplot(class_per_sample, aes(y = Abundance, x = reorder(Sample, SampleDate), fill = Class_plot))+
  facet_grid(~Year, scales = "free")+
  geom_col(position = "fill", stat="identity", size = 0)+ 
  scale_fill_manual(values = rev(pal))+
  ylab("Relative abundance")+
  theme(axis.text.x = element_blank(), axis.ticks = element_blank(), axis.title.x = element_blank())+
   scale_y_continuous(expand = c(0,0)) +
  labs(fill = "Bacterial class")+
   theme(plot.margin=unit(c(0.2,0.2,0.2,1),"cm"))+
  theme(legend.justification = c("left", "center"))



## TB stage
str(class_per_sample)
levels(class_per_sample$TB_stage)
levels(class_per_sample$TB_stage)<-c("TB unexposed", "TB exposed", "TB diseased")
#levels(class_per_sample$TB_stage_weight)<-c("TB unexposed", "TB exposed", "TB diseased")

table(class_per_sample$TB_stage)

plot_TB_sample<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = TB_stage))+
  geom_tile()+
  theme_void()+
scale_fill_manual(values = c( "forestgreen", "skyblue","brown2"))+ 
  theme(axis.title.y = element_text(angle = 0))+
  ylab("TB")+
  theme(legend.position="right", legend.direction="vertical", 
       legend.justification = c("left", "top"),
        legend.key.height = unit(0.3, 'cm'),
        legend.title=element_blank())


# temp max

pal_dichromat<-dichromat::colorschemes$BluetoOrange.12

pal_dichromat[13]<-"black"
pal_dichromat[14]<-"black"


plot_temp1<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = Temp_max))+geom_tile()+
  theme_void()+
 scale_fill_gradientn(colours = pal_dichromat)+
  theme(legend.position="right", legend.direction="vertical", 
         legend.justification = c("left", "top"),
        legend.key.height = unit(0.15, 'cm'),
        legend.title=element_blank()) +
  theme(axis.title.y = element_text(angle = 0))+
  ylab(" Max \ntemp")

## season

plot_season<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = Season))+
  geom_tile()+
  theme_void()+
  scale_fill_manual(values = c("gold", "forestgreen"))+
  theme(legend.position="right", legend.direction="vertical", 
        legend.justification = c("left", "top"),
        legend.key.height = unit(0.1, 'cm'),
        legend.title=element_blank())+
   theme(axis.title.y = element_text(angle = 0))+
  ylab("Season")





# year

year_names <- c(
                    `1997` = "",
                    `1998` = "",
                    `1999` = "1999",
                    `2000` = "",
                    `2001` = "2001",
                    `2002` = "2002",
                    `2003` = "2003",
                    `2004` = "2004",
                    `2005` = "2005",
                    `2006` = "2006",
                    `2007` = "2007",
                    `2008` = "2008",
                    `2009` = "2009",
                    `2010` = "2010",
                    `2011` = "2011",
                    `2012` = "2012",
                    `2013` = "2013",
                    `2014` = "2014",
                    `2015` = "2015",
                    `2016` = "2016",
                    `2017` = "2017",
                    `2018` = "2018",
                    `2019` = "")


# 13 x 6

# ####methods 
# ####methods 
# ####methods 
# ####methods 

drought_years<- c(1997, 1998, 2000, 2006, 2012, 2014, 2015, 2017, 2018)

class_per_sample$Drought<-ifelse(class_per_sample$Year %in% drought_years, "Drought", "Normal")

drought_cols<- c("red", "red", "skyblue", "red", "skyblue", # 1997 - 2001
                 "skyblue","skyblue","skyblue","skyblue", "red", # 2002 - 2006
                 "skyblue","skyblue","skyblue","skyblue","skyblue", # 2007 - 2011
                 "red", "skyblue", "red", "red", "skyblue", # 2012 - 2016
                 "red", "red" , "skyblue" ) # 2017 - 2019




plot_year<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = Drought))+
  geom_tile()+
  theme_void()+
 # scale_fill_viridis(discrete = F)+ 
  scale_fill_manual(values = c("red", "skyblue"))+
  facet_grid(~Year, scales = "free", space = "free_x",switch = "x", labeller = as_labeller(year_names))+

  theme(panel.spacing = unit(0.07, "lines"), 
         strip.background = element_blank(),
         strip.placement = "outside",
         strip.text.x = element_text(angle = 90, hjust = 0.5)) +
  
   xlab("")+
    theme(axis.title.y = element_text(angle = 0, hjust = ))+
  ylab("Year          ")+
  theme(axis.title.x = element_text(colour = "black"))+
  theme(plot.margin=unit(c(0,2.8,0.2,1.55),"cm")) +
    theme(legend.position="right", legend.direction="vertical", 
        legend.justification = c("left", "top"),
        legend.key.height = unit(0.2, 'cm'),
        legend.title=element_blank())



# rainfall

plot_rainfall<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = log10(sum_rainfall_2weeks+1)))+geom_tile()+
  theme_void()+
 scale_fill_gradientn(colours = pal_dichromat[1:12])+
   theme(legend.position="right", legend.direction="vertical", 
        legend.justification = c("left", "top"),
        legend.key.height = unit(0.2, 'cm'),
        legend.title=element_blank())+
   theme(axis.title.y = element_text(angle = 0))+
  ylab("Rainfall")

## body condition

scales::show_col(pal_dichromat_cond)

pal_dichromat_cond<-c("black", "black", "black", pal_dichromat[1:12])

pal_dichromat_cond[8]<-"white"
pal_dichromat_cond[9]<-"white"
pal_dichromat_cond[10]<-"white"

pal_dichromat_cond[16]<-"red"

plot_condition<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = Condition_weight))+
  geom_tile()+
  theme_void()+
 scale_fill_gradientn(colours = pal_dichromat_cond)+
   theme(legend.position="right", legend.direction="vertical", 
        legend.justification = c("left", "top"),
        legend.key.height = unit(0.2, 'cm'),
        legend.title=element_blank())+
   theme(axis.title.y = element_text(angle = 0))+
  ylab("Condition")


## without storage


sup_figabc<-ggarrange(plot_per_sample, plot_TB_sample, plot_condition, plot_temp1, plot_rainfall, plot_season, ncol = 1, align = "v", heights = c(8,1,1, 1,  1,1, 1, 1.5), labels= c("a)", "b)", "c)", "d)", "e)", "f)"))

ggarrange(sup_figabc, plot_year, ncol = 1, heights = c(7,1.1), labels = c(NA, "g)"))

# fig 2

ggsave("Figures/Fig2.pdf")
# Saving 16.4 x 7.97 in image

############## sup figure ################
############## sup figure ################
############## sup figure ################
############## sup figure ################

# social group

group_freq<-data.frame(table(class_per_sample$GroupAtSampling))

group_freq<-group_freq %>% arrange(-Freq)
head(group_freq, 10)

top_groups<-group_freq$Var1[1:15]

class_per_sample$GroupAtSampling_plot<-ifelse(class_per_sample$GroupAtSampling %in% top_groups, as.character(class_per_sample$GroupAtSampling), "Other")

mypal =  	pal_lancet("lanonc", alpha = 1)(7)
mypal5<- pal_rickandmorty("schwifty")(8)
mypal6<-brewer.pal(12,"Paired")
mypal7<-brewer.pal(8,"Dark2")
mypalx<-c(mypal6, mypal7)
scales::show_col(mypalx)


class_per_sample$GroupAtSampling_plot<-factor(class_per_sample$GroupAtSampling_plot)
levels(class_per_sample$GroupAtSampling_plot)

mypalx[9]<-"grey"



plot_group<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = GroupAtSampling_plot))+
  geom_tile()+
  theme_void()+
   scale_fill_manual(values = mypalx)+
   theme(axis.title.y = element_text(angle = 0))+
  ylab("Group")+
  labs(fill = "Group")+
  guides(fill=guide_legend(ncol=2))


# storage

plot_storage<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = Storage))+
  geom_tile()+
 theme_void()+
 # theme_bw()+
  scale_fill_manual(values = c("skyblue", "navyblue"), guide = guide_legend(reverse = TRUE))+ 
  theme(legend.position="right", legend.direction="vertical", 
         legend.key.height = unit(0.2, 'cm'),
        legend.justification = c("left", "top"), legend.title=element_blank())+
   theme(plot.margin=unit(c(0,0.2,0,0.2),"cm"))+
   theme(axis.title.y = element_text(angle = 0))+
  ylab("Storage")

# time

plot_time<-ggplot(class_per_sample, aes(y = 1, x = reorder(Sample, SampleDate), fill = HoursAfterSunrise))+geom_tile()+
  theme_void()+
  theme_bw()+
 scale_fill_viridis_c(option = "plasma", direction= -1)+
   theme(legend.position="right", legend.direction="horizontal", 
        legend.justification = c("left", "top"),
        legend.key.height = unit(0.2, 'cm'),
        legend.title=element_blank())+
   theme(axis.title.y = element_text(angle = 0))+
  ylab("Time \nof day")

social_group<-ggarrange(plot_per_sample,  plot_group,plot_time, plot_storage, ncol = 1, align = "v", heights = c(8,1,1, 1), labels= c("a)", "b)", "c)", "d)"), legend = F)

social_group

mb_legend<-get_legend(plot_per_sample)
plot(mb_legend)
group_legend<-get_legend(plot_group)
plot(group_legend)
storage_legend<-get_legend(plot_storage)
plot(storage_legend)
time_legend<-get_legend(plot_time)
plot(time_legend)

```

# 3. Fig S3. Visualising raw data per individual

```{r}


####### microbiome plot

meerkat_merged_ID<-merge_samples(meerkat_filtered,"IndividID", fun=mean)
sample_data(meerkat_merged_ID)$IndividID<-sample_names(meerkat_merged_ID)
meerkat_merged_ID_class<-tax_glom(meerkat_merged_ID, taxrank = "Class")
sample_data(meerkat_merged_ID_class)<-sample_data(meerkat_merged_ID_class)[,1]
meerkat_merged_ID_class<-microbiome::transform(meerkat_merged_ID_class, "compositional") 
class_per_ID<-psmelt(meerkat_merged_ID_class)

class_abundance<-class_per_ID %>% group_by(Class) %>% summarise(mean = mean(Abundance))
class_abundance<-class_abundance[order(-class_abundance$mean),]
top_class<-class_abundance$Class[1:7]

class_per_ID$Class_plot<-as.character(ifelse(class_per_ID$Class %in% top_class, class_per_ID$Class, "Other"))


class_per_ID$Class_plot<-factor(class_per_ID$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))

## order by birthyear #####
## order by birthyear #####
## order by birthyear #####
## order by birthyear #####

metadata<-data.frame(sample_data(meerkat_filtered))
metadata$BirthYear<-lubridate::year(sample_data(meerkat_filtered)$BirthDate)
metadata<-metadata[,c("feature.id", "IndividID", "TB_status", "BirthYear","BirthDate")]
head(metadata)
metadata2<-distinct(metadata, IndividID, .keep_all = T)
metadata2<-metadata2 %>% arrange(BirthDate)
head(metadata2)

ID_order<-metadata2$IndividID

dim(class_per_ID)

class_per_ID<-merge(class_per_ID, metadata2, by.y = "IndividID", by.x = "Sample")

class_per_ID$TB_status<-factor(class_per_ID$TB_status, levels = c("unexposed", "exposed_asymptomatic", "exposed_symptomatic"))

levels(class_per_ID$TB_status)<-c("TB unexposed", "TB exposed", "TB diseased")

metadata2$IndividID<-factor(metadata2$IndividID, levels =ID_order)
class_per_ID$IndividID<-factor(class_per_ID$IndividID, levels =ID_order)

head(class_per_ID)

#class_per_ID<-class_per_ID%>% group_by(IndividID, Class_plot)%>%summarize(Abundance = sum(Abundance))
class_per_ID<-data.frame(class_per_ID)

#########################
#########################
#########################
#########################

microbiome_barplot <- ggplot(class_per_ID, aes(x = IndividID, y = Abundance, fill = Class_plot, col = Class_plot))+
  theme_gray(base_size = 14)+
  geom_bar( stat="identity", width = 0.5)+ 
  scale_fill_manual(values = rev(pal))+
  scale_color_manual(values = rev(pal), guide = "none")+
  theme(legend.position = "right")+
  theme(axis.text.x = element_blank(), axis.title.x = element_blank(), axis.ticks.x = element_blank())+
   theme(plot.margin=unit(c(0.4,0,0,0),"cm"))+
     scale_y_continuous(expand = c(0,0)) +
  labs(fill = "Bacterial class")+
  theme(legend.justification = c("left", "center"))


######## TB 

p2<-ggplot(metadata2, aes(x = IndividID, y = 1, fill = TB_status))+
  geom_tile(col = "white")+
  theme_void(base_size = 14)+
  scale_fill_manual(values = c("forestgreen", "cyan3", "brown2"))+
  theme(legend.position = "right", legend.justification = "left") +
  theme(axis.text.x = element_blank(), axis.title.x = element_blank(), axis.ticks.x = element_blank())+
  theme(axis.text.y = element_blank())+
      theme(axis.title.y = element_text(angle = 90, vjust=0.5))+
  ylab("TB \nstatus")+
   theme(plot.margin=unit(c(0,0,0,0.1),"cm"))+
  labs(fill = "TB status")



######## birth year




p3<-ggplot(metadata2, aes(y = 1, x = IndividID, fill = BirthYear))+
  geom_tile()+
  theme_void(base_size = 14)+
  scale_fill_viridis(discrete = F)+ 
  facet_grid(~BirthYear, scales = "free", space = "free_x",switch = "x")+

  theme(panel.spacing = unit(0.05, "lines"), 
         strip.background = element_blank(),
         strip.placement = "outside",
         strip.text.x = element_text(angle = 90, size = , hjust = 0.5)) +
  
   xlab("")+
    theme(axis.title.y = element_text(angle = 90, vjust=9))+
  ylab("Birth \nyear")+
  theme(axis.title.x = element_text(colour = "black"))+
   theme(plot.margin=unit(c(0, 5.7, 0.2,1.1),"cm"))+
  theme(legend.position="none") 



figsxab<-ggarrange(microbiome_barplot,p2, ncol = 1, align = "v", heights = c(2,0.5))

ggarrange(figsxab, p3, ncol = 1, heights = c(3,0.8))

ggsave("Figures/FigS6.pdf")
# Saving 16.9 x 7.97 in image

fig_ID<-ggarrange(figsxab, p3, ncol = 1, heights = c(3,0.8))






####################


bacteroidia_2005<-subset(class_per_ID, Class_plot == "Bacteroidia" & BirthYear <2006 )
bacteroidia_2014<-subset(class_per_ID, Class_plot == "Bacteroidia"& BirthYear >2013 )

length(unique(bacteroidia_2005$IndividID))
length(unique(bacteroidia_2014$IndividID))

summary(bacteroidia_2005$Abundance)
sd(bacteroidia_2005$Abundance)
summary(bacteroidia_2014$Abundance)
sd(bacteroidia_2014$Abundance)



bacilli_2005<-subset(class_per_ID, Class_plot == "Bacilli" & BirthYear <2006 )
bacilli_2014<-subset(class_per_ID, Class_plot == "Bacilli"& BirthYear >2014 )

summary(bacilli_2005$Abundance)
sd(bacilli_2005$Abundance)
summary(bacilli_2014$Abundance)

sd(bacilli_2014$Abundance)

```

# 6. Figs S4, S5 and S9. Split by storage, season and group

```{r}


storage_ps_class<-tax_glom(meerkat_filtered, taxrank = "Class")

storage_ps_class<-core(storage_ps_class, detection = 10, prevalence = 0)

sample_data(storage_ps_class)<-sample_data(storage_ps_class)[,c("feature.id", "Storage", "IndividID", "Year")]
class_per_storage<-psmelt(storage_ps_class)

class_per_storage$Class_plot<-as.character(ifelse(class_per_storage$Class %in% top_class, class_per_storage$Class, "Other"))



class_per_storage$Class_plot<-factor(class_per_storage$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))


class_per_storage$Storage<-factor(class_per_storage$Storage, levels = c("FROZEN", "FREEZEDRIED"))

ggplot(class_per_storage, aes(x =reorder(Sample, Year), y = Abundance, fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
 # coord_flip()+
  scale_fill_manual(values = rev(pal))+
  facet_wrap(~Storage, scales = "free")

################

class_per_storage$Storage<-factor(class_per_storage$Storage, levels = c("FROZEN", "FREEZEDRIED"))

ggplot(class_per_storage, aes(x =Year, y = Abundance, fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
  ylab("Relative abundance")+
 facet_wrap(~Storage, ncol = 1, strip.position="right")+
  scale_fill_manual(values = rev(pal))+
  xlim(c(1996, 2020))+
 # theme(plot.margin=unit(c(0.2,0,0,0.1),"cm"))+
  labs(fill = "Bacterial class")+
  ggtitle("Yearly composition by storage method")


```

-   Season

```{r}


season_ps_class<-tax_glom(meerkat_filtered, taxrank = "Class")

season_ps_class<-core(season_ps_class, detection = 10, prevalence = 0)

sample_data(season_ps_class)<-sample_data(season_ps_class)[,c("feature.id", "Season", "IndividID", "Year")]
class_per_season<-psmelt(season_ps_class)

class_per_season$Class_plot<-as.character(ifelse(class_per_season$Class %in% top_class, class_per_season$Class, "Other"))



class_per_season$Class_plot<-factor(class_per_season$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))



ggplot(class_per_season, aes(x =reorder(Sample, Year), y = Abundance, fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
 # coord_flip()+
  scale_fill_manual(values = rev(pal))+
  facet_wrap(~Season, scales = "free")

################


ggplot(class_per_season, aes(x =Year, y = Abundance, fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
  ylab("Relative abundance")+
 facet_wrap(~Season, ncol = 1, strip.position="right")+
  scale_fill_manual(values = rev(pal))+
  xlim(c(1996, 2020))+
 # theme(plot.margin=unit(c(0.2,0,0,0.1),"cm"))+
  labs(fill = "Bacterial class")+
  ggtitle("Yearly composition by season")


```

- TB group

```{r}


TBgroup_ps_class<-tax_glom(meerkat_filtered, taxrank = "Class")

TBgroup_ps_class<-core(TBgroup_ps_class, detection = 10, prevalence = 0)

sample_data(TBgroup_ps_class)<-sample_data(TBgroup_ps_class)[,c("feature.id", "TB_present_group", "IndividID", "Year", "SampleDate")]
class_per_TBgroup<-psmelt(TBgroup_ps_class)

class_per_TBgroup$Class_plot<-as.character(ifelse(class_per_TBgroup$Class %in% top_class, class_per_TBgroup$Class, "Other"))



class_per_TBgroup$Class_plot<-factor(class_per_TBgroup$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))

class_per_TBgroup<-subset(class_per_TBgroup, !is.na(TB_present_group))



ggplot(class_per_TBgroup, aes(x =reorder(Sample, SampleDate), y = Abundance, fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
 # coord_flip()+
  scale_fill_manual(values = rev(pal))+
  facet_wrap(~TB_present_group, scales = "free")

################


ggplot(class_per_TBgroup, aes(x =Year, y = Abundance, fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
  ylab("Relative abundance")+
 facet_wrap(~TB_present_group, ncol = 1, strip.position="right")+
  scale_fill_manual(values = rev(pal))+
  xlim(c(1996, 2020))+
 # theme(plot.margin=unit(c(0.2,0,0,0.1),"cm"))+
  labs(fill = "Bacterial class")+
  ggtitle("Yearly composition by whether TB disease in social group")


```

-   Social groups

```{r}

metadata_groups<-data.frame(sample_data(meerkat_filtered))

group_freq<-data.frame(table(metadata_groups$GroupAtSampling))

group_freq<-group_freq %>% arrange(-Freq)
head(group_freq, 10)

top_groups<-group_freq$Var1[1:15]


group_ps_class<-tax_glom(meerkat_filtered, taxrank = "Class")

group_ps_class<-core(group_ps_class, detection = 10, prevalence = 0)

sample_data(group_ps_class)<-sample_data(group_ps_class)[,c("feature.id", "Storage", "IndividID", "Year", "SampleDate", "GroupAtSampling")]
class_per_storage<-psmelt(group_ps_class)

class_per_storage$Class_plot<-as.character(ifelse(class_per_storage$Class %in% top_class, class_per_storage$Class, "Other"))



class_per_storage$Class_plot<-factor(class_per_storage$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))

class_per_storage$Group_plot<-ifelse(class_per_storage$GroupAtSampling %in% top_groups, as.character(class_per_storage$GroupAtSampling), "Other")



class_per_storage$Group_plot<-forcats::fct_reorder(class_per_storage$Group_plot, class_per_storage$Year, mean)

length(unique(subset(class_per_storage, Group_plot != "Other")$feature.id))

ggplot(subset(class_per_storage, Group_plot != "Other"), aes(x =Year, y = Abundance, fill = Class_plot))+
  geom_col(position = "fill", stat="identity", size = 0)+ 
  ylab("Realtive abundance")+
  scale_fill_manual(values = rev(pal))+
  facet_wrap(~Group_plot, ncol = 1, strip.position="right")+
  ggtitle("Yearly composition by major social groups")+
  scale_y_continuous(breaks = c(0.2, 0.4, 0.6, 0.8))+
  labs(fill = "Bacterial class")




```

# Fig 3: Microbiome composition by year barplot

```{r}


meerkat_merged_year<-merge_samples(meerkat_filtered,"Year", fun=mean)

sample_data(meerkat_merged_year)$Year<-sample_names(meerkat_merged_year)

meerkat_merged_year_class<-tax_glom(meerkat_merged_year, taxrank = "Class")

sample_data(meerkat_merged_year_class)<-sample_data(meerkat_merged_year_class)[,1]

#meerkat_merged_year_class<-microbiome::transform(meerkat_merged_year_class, "compositional") 
class_per_year<-psmelt(meerkat_merged_year_class)



class_per_year$Class_plot<-as.character(ifelse(class_per_year$Class %in% top_class, class_per_year$Class, "Other"))

class_per_year$Class_plot<-factor(class_per_year$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))

table(class_per_year$Sample)

str(class_per_year)
class_per_year$Sample<- as.integer(class_per_year$Sample)

class_per_year<-class_per_year%>% group_by(Sample, Class_plot)%>%summarize(Abundance = sum(Abundance))
class_per_year<-data.frame(class_per_year)
class_per_year$Sample<-factor(class_per_year$Sample)


plot_mb_year<-ggplot(class_per_year, aes(x =Sample, y = Abundance, fill = Class_plot))+
  geom_bar( stat="identity",  size = 0, position = "fill")+ 
   theme_bw()+
  scale_fill_manual(values = rev(pal))+
  theme(legend.position = "right")+
  xlab("")+ 
theme(axis.text.x = element_text(angle = 90, vjust = 0.9, hjust=1))+
  labs(fill = "Bacterial class")+
 #  theme(plot.margin=unit(c(0.2,0.2,0,0.2),"cm"))+
   scale_y_continuous(expand = c(0,0))+
  # scale_x_continuous(breaks =  c(1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 
   #                             2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015,
  #                              2016, 2017, 2018, 2019), expand = c(0,0))+
  xlab("Year")


year_freq<-data.frame(table(sample_data(meerkat_filtered)$Year))


head(year_freq)
  str(year_freq)

year_hist<-ggplot(year_freq, aes(x = Var1, y = Freq)) + geom_col(fill = "lightgray", col = "black")+
# theme(plot.margin=unit(c(0.2,5,0.2,0.2),"cm"))+
  theme(axis.text.x = element_blank(), axis.title.x = element_blank())+
  theme(legend.position = "none")+
  scale_y_sqrt(breaks = c(0,20,50,100,200))+
  geom_hline(yintercept = 20, col = "black", linetype = "longdash")

year_plot<-ggarrange(year_hist, plot_mb_year, ncol = 1, align = "v", heights = c(1,7), legend = "none")+
     theme(plot.margin=unit(c(0.2,0.2,0.2,0.6),"cm"))




## tb

TB_summary_year2<-subset(TB_summary_year, TB_status == "TB diseased")

subset(TB_summary_year, Year == 1997)
str(TB_summary_year2)

extra_rows<-data.frame(as.integer(1997), as.factor("TB diseased")   , as.integer(0) ,  as.integer(165) ,as.numeric(0) ,as.numeric(0))
names(extra_rows)<-names(TB_summary_year2)

extra_rows2<-data.frame(as.integer(1998), as.factor("TB diseased")   , as.integer(0) ,  as.integer(165) ,as.numeric(0) ,as.numeric(0))
names(extra_rows2)<-names(TB_summary_year2)


TB_summary_year3<-rbind(TB_summary_year2, extra_rows, extra_rows2)
TB_summary_year3$Year<-factor(TB_summary_year3$Year)
summary(TB_summary_year3)


TB_hist<-ggplot(TB_summary_year3, aes(y = Proportion, x = Year))+
  geom_col(fill = "brown2", col = "black")+
# theme(plot.margin=unit(c(0.2,5,0.2,0.2),"cm"))+
  theme(axis.text.x = element_blank(), axis.title.x = element_blank())+
  theme(legend.position = "none")

year_plot<-ggarrange(year_hist, TB_hist, plot_mb_year, ncol = 1, align = "v", heights = c(1,1, 5), legend = "none", labels = c("b)", "c)", "d)"), label.x = -0.07)+
     theme(plot.margin=unit(c(0.2,0.2,0.2,1),"cm"))


```

# Alpha diversity

```{r}

summary(sample_data(meerkat_filtered)$Observed)


sample_data(meerkat_filtered)$JuliDate<-as.numeric(sample_data(meerkat_filtered)$SampleDate - min(sample_data(meerkat_filtered)$SampleDate))

metadata<-data.frame(sample_data(meerkat_filtered))
str(metadata)

m_alpha <- mgcv::bam(Observed~
   # s(JuliDate, bs = "cr")+
    s(HoursAfterSunrise, bs = "cr")+
    s(month, bs = "cc")+
    s(AgeAtSamplingYears, bs = "cr")+
    s(Seq_depth_nospikein, bs = "cr")+
    Storage+
      TB_stage+
    Seq_run+
    HoursTilFrozen+
    s(IndividID, bs = "re")+
    s(GroupAtSampling, bs = "re"),
  #correlation = corARMA(form = ~ 1|Year, p = 3),
  data=metadata)
summary(m_alpha)


gratia::draw(m_alpha, select = c(1:3), ncol = 3, scales = "fixed")

ggplot(metadata, aes(x = as.factor(Year), y = Observed))+geom_boxplot()+
  geom_jitter(width = 0.1, alpha = 0.3)

ggplot(metadata, aes(x = JuliDate, y = Observed))+
  geom_point(alpha = 0.3)


```

# 7. Constrained ordination of gut microbiota

-   Constrained ordination using CAP
-   Also test all supertypes

```{r ordination}

scaled_core<-microbiome::core(meerkat_filtered, detection = 100, prevalence = 0.01)

scaled_core<-tax_glom(meerkat_filtered, taxrank = "Genus")

scaled_core<-microbiome::core(scaled_core, detection = 10, prevalence = 0.005)


scaled_core


```

-   organise variables

```{r}


# otutable

otutable<-data.frame(t(scaled_core@otu_table@.Data))
dim(otutable)
otutable[1:5,1:5]
colnames(otutable)<-taxa_names(scaled_core)

metadata<-data.frame(sample_data(scaled_core))
names(metadata)


metadata<-metadata[,c( "TB_stage","TB_status", "TB_survival",
                      "Condition_weight", "Season","Year",
                      "IndividID", "TimeCat","HoursAfterSunrise",  "Temp_max",  "sum_rainfall_month",
                      "Storage", "Seq_depth_nospikein", "Seq_run")]


## TB variables

TB_stage<-metadata$TB_stage
TB_status<-metadata$TB_status
TB_survival<-metadata$TB_survival



# condition and season

summary(metadata$Condition_weight)
metadata$Condition_cat <- ifelse(metadata$Condition_weight < summary(metadata$Condition_weight)[2], "BAD", "AVERAGE")
metadata$Condition_cat <- ifelse(metadata$Condition_weight > summary(metadata$Condition_weight)[5], "GOOD", metadata$Condition_cat)

condition<-metadata$Condition_cat
season <- metadata$Season

## rainfall

summary(log10(metadata$sum_rainfall_month+1))
metadata$Rainfall_cat <- ifelse(metadata$sum_rainfall_month == summary(metadata$sum_rainfall_month)[2], "LOW", "AVERAGE")
metadata$Rainfall_cat <- ifelse(metadata$sum_rainfall_month > summary(metadata$sum_rainfall_month)[5], "HIGH", metadata$Rainfall_cat)
table(metadata$Rainfall_cat)

# temp
hist(metadata$Temp_max)
summary(metadata$Temp_max)
metadata$Max_temp_cat <- ifelse(metadata$Temp_max < summary(metadata$Temp_max)[2], "LOW", "AVERAGE")
metadata$Max_temp_cat <- ifelse(metadata$Temp_max > summary(metadata$Temp_max)[5], "HIGH", metadata$Max_temp_cat)
table(metadata$Max_temp_cat)


## max temp

max_temp<-metadata$Max_temp_cat
#hist(max_temp)

# year

hist(metadata$Year)

metadata<-metadata %>% mutate(YearCat = case_when((Year <2005 ~ "Pre-2005"),
                                                  (Year >=2005 & Year < 2010) ~ "2005-2010",
                                                  (Year >=2010 & Year < 2015) ~ "2010-2015",
                                                  (Year >=2015) ~ "2015-2019"))
                                                  
  

YearCat<-as.character(metadata$YearCat)




### control variables

IndividID<-metadata$IndividID
time<-metadata$TimeCat
#time<-as.numeric(scale(metadata$HoursAfterSunrise))
maxtemp<-metadata$Max_temp_cat
samplingtemp<-metadata$HourlyTemp
foraging<-log10(metadata$HoursSinceForagingStart+0.1)
rainfall<-metadata$Rainfall_cat
# methods

storage<-metadata$Storage
seq_depth<-as.numeric(scale(metadata$Seq_depth_nospikein))
seq_run<-metadata$Seq_run

```

-   Generate distance matrices

```{r}
###### genus level 

wunifrac_dist<-distance(scaled_core, method = "wunifrac")
#saveRDS(wunifrac_dist, "wunifrac_genus_ordination.RDS")

unifrac_dist<-distance(scaled_core, method = "unifrac")
#saveRDS(wunifrac_dist, "unifrac_genus_ordination.RDS")

bray_dist<-distance(scaled_core, method = "bray")
#saveRDS(bray_dist, "bray_genus_ordination.RDS")


# asv level ###
# asv level ###
# asv level ###
# asv level ###

scaled_core_asv<-microbiome::core(meerkat_filtered, detection = 100, prevalence = 0.005)

wunifrac_dist_asv<-distance(scaled_core_asv, method = "wunifrac")
#saveRDS(wunifrac_dist_asv, "wunifrac_asv_ordination.RDS")

unifrac_dist_asv<-distance(scaled_core_asv, method = "unifrac")
#saveRDS(wunifrac_dist_asv, "unifrac_asv_ordination.RDS")

bray_dist_asv<-distance(scaled_core_asv, method = "bray")
#saveRDS(bray_dist_asv, "bray_asv_ordination.RDS")



```

#### Apply models

```{r}

#Calculate distance based RDA using Weighted UniFrac distance matrix 
#Calculate distance based RDA using Weighted UniFrac distance matrix 
#Calculate distance based RDA using Weighted UniFrac distance matrix 
#Calculate distance based RDA using Weighted UniFrac distance matrix 



final_model<-capscale(wunifrac_dist ~ 
                       time + 
                        TB_stage+
                        condition+
                        
                        rainfall+
                        
                       YearCat+
                      
                        storage+
                        seq_depth+
                       IndividID, 
                env = metadata, comm = otutable) 

ggord(final_model,  
      xlims = c(min(data.frame(final_model$CCA$wa)$CAP1)-1, 
                max(data.frame(final_model$CCA$wa)$CAP1)+1),
      ylims = c(min(data.frame(final_model$CCA$wa)$CAP2)-1, 
                max(data.frame(final_model$CCA$wa)$CAP2)+1), 
      addsize = 3, size = 1, txt = 2)


RsquareAdj(final_model)

unifrac_model<-capscale(unifrac_dist ~ 
                       time + 
                        TB_stage+
                        condition+
                         rainfall+
                       #  maxtemp+
                       YearCat+
                        storage+
                        seq_depth+
                       IndividID, 
                env = metadata, comm = otutable) 


bray_model<-capscale(bray_dist ~ 
                       time + 
                        TB_stage+
                        condition+
                          rainfall+
                       #maxtemp+
                       YearCat+
                        storage+
                        seq_depth+
                       IndividID, 
                env = metadata, comm = otutable) 
ggord(bray_model,  
      xlims = c(min(data.frame(final_model$CCA$wa)$CAP1)-1, 
                max(data.frame(final_model$CCA$wa)$CAP1)+1),
      ylims = c(min(data.frame(final_model$CCA$wa)$CAP2)-1, 
                max(data.frame(final_model$CCA$wa)$CAP2)+1), 
      addsize = 3, size = 1, txt = 2)



```

```{r eval = F}
##### statistical test 
##### statistical test 
##### statistical test 

# weighted unifrac

final_model_anova<-anova.cca(final_model, by="terms")

final_model_anova

final_model_anova_df<-data.frame(final_model_anova)
# write.csv(final_model_anova_df, "dbrda_wunifrac_summary_genus.csv")


## unifrac ##

unifrac_anova<-anova.cca(unifrac_model, by="terms")

unifrac_anova

unifrac_anova_df<-data.frame(unifrac_anova)
#write.csv(unifrac_anova_df, "dbrda_unifrac_summary_genus.csv")


# bray


bray_anova<-anova.cca(bray_model, by="terms")

bray_anova

bray_anova_df<-data.frame(bray_anova)
#write.csv(bray_anova_df, "dbrda_bray_summary_genus.csv")
```

### Get dominant classes per sample

```{r}


##dominant taxa

#scaled_core_genus<- tax_glom(scaled_core, taxrank = "Genus")

dominant_phylo<-scaled_core
#dominant_phylo<-scaled_core_genus
sample_data(dominant_phylo)<-sample_data(dominant_phylo)[,"feature.id"]

# melt data
psmelt_df<-speedyseq::psmelt(dominant_phylo)
head(psmelt_df)

dominant_taxa_df<-psmelt_df %>% 
  group_by(feature.id) %>% 
  arrange(feature.id, -Abundance) %>% 
  dplyr::mutate(rank=rank(-Abundance))

head(dominant_taxa_df)

dominant_taxa_df<-subset(dominant_taxa_df, rank ==1)
dominant_taxa_df$Class<-factor(dominant_taxa_df$Class)

# add to metadata 
sample_data(scaled_core)$DomClass <-vlookup(sample_data(scaled_core)$feature.id, dominant_taxa_df, lookup_column = "Sample", result_column = "Class")

top_class

sample_data(scaled_core)$DomClass_plot<-ifelse(sample_data(scaled_core)$DomClass %in% top_class, as.character(sample_data(scaled_core)$DomClass), "Other")
unique(sample_data(scaled_core)$DomClass_plot)

#sample_data(scaled_core)$DomClass_plot<-factor(sample_data(scaled_core)$DomClass_plot, levels =c(top_class, "Other"))

sample_data(scaled_core)$DomClass_plot<-factor(sample_data(scaled_core)$DomClass_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))

pal<-brewer.pal(10,"Paired")
pal<-c(pal, "darkgrey")
scales::show_col(pal)
pal<-pal[c(2,1,11,4,3,6,9,10)]
pal<-rev(pal)


```

### Oridination figure

```{r}

######################################

## extract data from model

final_model_df<-scores(final_model)
str(final_model_df)


# extract CAP scores

vectors_df<-data.frame(final_model_df$sites)

vectors_df$feature.id<-row.names(vectors_df)

# merge with info on dominant family

sample_metadata<-data.frame(sample_data(scaled_core))[,c("feature.id", "DomClass_plot")]
vectors_df<-merge(vectors_df, sample_metadata, by = "feature.id")

# basic plot


ggplot(vectors_df, aes(x = CAP1, y = CAP2))+
  geom_point(aes(fill = DomClass_plot), pch = 21, size = 4)+
  scale_fill_manual(values = rev(pal))+
  theme_bw()

#### add arrows #########
#### add arrows #########
#### add arrows #########
#### add arrows #########

centroids_df<-data.frame(final_model_df$centroids)
centroids_df<-centroids_df[c(1:5,7:8,10:15),]
#centroids_df<-centroids_df[c(1,2,4,6:10,12:15),]
centroids_df$Label<-c("Afternoon", "Morning",  "TB unexposed", "TB exposed", "TB clinical signs", "Bad condition", "Good condition","High rainfall", "Low rainfall" , "2005-2010", "2010-2015", "2015-2019", "Pre-2005")

# scale so they are more plottable


centroids_df[1,1] <-centroids_df[1,1] *0.7
centroids_df[1,2] <-centroids_df[1,2] *0.7

centroids_df[13,1] <-centroids_df[13,1] *0.7
centroids_df[13,2] <-centroids_df[13,2] *0.7

centroids_df$CAP1<-centroids_df$CAP1 * 1.5
centroids_df$CAP2<-centroids_df$CAP2 * 1.5

ggplot(vectors_df, aes(x = CAP1, y = CAP2))+
  geom_point(aes(fill = DomClass_plot), pch = 21, size = 4)+
  scale_fill_manual(values = rev(pal))+
  theme_bw()+
  
  # add arrows
  
  geom_segment(data=centroids_df, aes(x = 0, y = 0, xend = CAP1*3, yend = CAP2*3),
    arrow = arrow(length = unit(0.5, "cm"), type = "closed"), lwd = 1,  col = "black")+
  ggrepel::geom_label_repel(data=centroids_df,  
    alpha = 0.9, col = "black", size = 3, fill = "yellow", 
    #hjust = c(0,1), 
    #vjust = c(1,1),
    aes(CAP1 *3, CAP2*3, label = Label))

### add species scores ###
### add species scores ###
### add species scores ###
### add species scores ###


species_scores<-data.frame(final_model_df$species)

hist(species_scores$CAP1, breaks = 50)
hist(species_scores$CAP2, breaks = 50)

summary(species_scores$CAP1)
summary(species_scores$CAP2)


species_scores<-subset(species_scores, CAP1 < -1  | CAP2 > 0.2 | CAP2 < -0.2)

#species_scores$CAP1<-species_scores$CAP1*0.8
#species_scores$CAP2<-species_scores$CAP2*0.8



species_scores$ASV<-row.names(species_scores)

taxonomy<-data.frame(tax_table(scaled_core))
head(taxonomy)
taxonomy$ASV<-row.names(taxonomy)

species_scores<-merge(species_scores, taxonomy, by = "ASV")

subset(species_scores, Genus == "Lactococcus")

#################

species_scores[4,2]<-species_scores[4,2]*0.2
species_scores[4,3]<-species_scores[4,3]*0.2


species_scores[5,2]<-species_scores[5,2]*0.8
species_scores[5,3]<-species_scores[5,3]*0.8


species_scores<-species_scores[-6,]

species_scores<-subset(species_scores, 
                       Genus == "Clostridium sensu stricto 1" |
                         Genus == "Blautia" |
                         Genus == "Fusobacterium" |
                         Genus == "Bacteroides" |
                         Genus == "Phascolarctobacterium" |
                         Genus == "Lactococcus" |
                         Genus == "Paeniclostridium" |
                         Genus == "[Ruminococcus] torques group")

species_scores[6,3]<-species_scores[6,3]*0.8
species_scores[6,2]<-species_scores[6,2]*0.8

###################
###################
###################
###################


ord_plot<-ggplot(vectors_df, aes(x = CAP1, y = CAP2))+
  geom_point(aes(fill = DomClass_plot), pch = 21, size = 3)+
  scale_fill_manual(values = rev(pal))+
  theme_bw()+
  xlab("CAP1 (40 %)")+ylab("CAP2 (14%)")+
 guides(fill = guide_legend(override.aes = list(size = 7)))+
  labs(fill = "Bacterial class")+
  
  # add arrows
  
  geom_segment(data=centroids_df, aes(x = 0, y = 0, xend = CAP1*3, yend = CAP2*3),
    arrow = arrow(length = unit(0.5, "cm"), type = "closed"), lwd = 1,  col = "black")+
  ggrepel::geom_label_repel(data=centroids_df,  
    alpha = 0.9, col = "black", size = 3, fill = "yellow", 
    aes(CAP1 *3, CAP2*3, label = Label))+
  
  # add species
  geom_point(data = species_scores,  size = 4, col = "black", pch = 8, stroke = 0, alpha = 0)+
   ggrepel::geom_label_repel(data=species_scores,  
    alpha = 0.9, col = "black", size = 2.5, fill = "skyblue", 
    aes( label = Genus))



ggarrange(ord_plot,year_plot,  ncol = 2 , common.legend = T, legend = "right", widths = c(3.5,2), labels = c("a)", NA))

ggsave("Figures/Fig3.pdf")


#Saving 11.6 x 5.91 in image

```

#### Add captive meerkats
```{r}
########################## add captive meerkats
########################## add captive meerkats
########################## add captive meerkats

pilot<- readRDS("C:\\Users\\risel\\Dropbox\\Sommer postdoc\\Meerkat project\\PROJECTS\\MeerkatSpikeInData\\Analysis updated\\1. TEMPORAL DYNAMICS ANALYSIS\\1_PUBLISHED DATASET AND CODE\\pilot_study_data_phyloseq.RDS")

pilot

pilot_genus<-tax_glom(pilot, taxrank = "Genus")
pilot_class<-tax_glom(pilot, taxrank = "Class")

#pilot_clr <- microbiome::transform(pilot_class, "clr")
pilot_comp <- microbiome::transform(pilot_class, "compositional")


sample_data(pilot_comp)<-sample_data(pilot_comp)[,c("feature.id")]

class_per_sample<-psmelt(pilot_comp)

head(class_per_sample)

top_class<-c("Bacteroidia","Fusobacteriia" ,  "Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli")


class_per_sample$Class_plot<-as.character(ifelse(class_per_sample$Class %in% top_class, class_per_sample$Class, "Other"))

class_per_sample %>% group_by(Class) %>% summarize(mean = mean(Abundance)) %>% arrange(-mean)



class_per_sample$Class_plot<-factor(class_per_sample$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli"))

head(class_per_sample)

metadata<-data.frame(sample_data(pilot))
names(metadata)

metadata<-metadata[,c("feature.id", "Storage", "IndividID")]

class_per_sample<- merge(class_per_sample, metadata, by = "feature.id")

head(class_per_sample)

captive_meerkats<-ggplot(subset(class_per_sample, Storage == "FROZEN"), aes(y = Abundance, x = IndividID, fill = Class_plot))+
  #facet_grid(~IndividID, scales = "free_x")+
  geom_col(position = "fill", stat="identity")+ 
  scale_fill_manual(values = rev(pal))+
  ylab("Relative abundance")+
   scale_y_continuous(expand = c(0,0)) +
  labs(fill = "Bacterial class")+
   theme(plot.margin=unit(c(2.5,0.2,0.6,0.2),"cm"))+
  theme(legend.justification = c("left", "center"))+
#  theme(axis.title.x = element_blank())+ #axis.text.x = element_text(angle = 90, hjust = 1))+
  theme(axis.title.y = element_blank(), axis.text.y = element_blank())+
  xlab("ID")


#lower<-ggarrange(plot_mb_year, captive_meerkats, ncol = 2 , common.legend = T, legend = "right", widths = c(2,1), labels = c("b)", "c)"), align = "h")

#ggarrange(ord_plot, lower, ncol = 1, heights = c(3,1))



```

# 8. Model genus level changes

Make an 'other' category for rare taxa!

```{r}

genus_ps<-tax_glom(meerkat_filtered, taxrank = "Genus")
#genus_ps<-tax_glom(subset_samples(meerkat_filtered, Storage == "FREEZEDRIED"), taxrank = "Genus")

genus_abundances_df<-data.frame(taxa_sums(genus_ps))

genus_abundances_df$ASV<-row.names(genus_abundances_df)

taxonomy<-data.frame(tax_table(genus_ps))
head(taxonomy)
taxonomy$ASV<-row.names(taxonomy)


genus_abundances_df<-merge(genus_abundances_df, taxonomy, by = "ASV")
names(genus_abundances_df)[2]<-"Abundance"

genus_abundances_df<-genus_abundances_df %>% arrange(-Abundance)

top_genera<- genus_abundances_df$ASV[1:30]

#top_genus_ps<-prune_taxa(top_genera, genus_ps)
#sample_data(top_genus_ps)<-sample_data(top_genus_ps)[,c( "feature.id")]


sample_data(genus_ps)<-sample_data(genus_ps)[,c( "feature.id")]

top_genera_melt<-psmelt(genus_ps)
head(top_genera_melt)

unique(top_genera_melt$Genus)

# make a category for rare genera

top_genera_melt$Genus2<-ifelse(top_genera_melt$OTU %in% top_genera, top_genera_melt$Genus, "Rare taxa")

head(top_genera_melt)

top_genera_melt2<-top_genera_melt %>% group_by(Sample, Genus2) %>% summarize(Abundance2 = sum(Abundance))

head(top_genera_melt2)

metadata<-data.frame(sample_data(meerkat_filtered))
#metadata<-data.frame(sample_data(subset_samples(meerkat_filtered, Storage == "FREEZEDRIED")))

top_genera_melt3<- merge(top_genera_melt2, metadata, by.x = "Sample", by.y = "feature.id")
names(top_genera_melt3)
head(top_genera_melt3)


```

-   Year effects

```{r}

# loop to fit GAMMs to each genus

year_estimates<-list()

taxanames<-unique(top_genera_melt2$Genus2)

for (i in 1:length(taxanames)){
  tryCatch({ #catch errors
    print(i)
  
    taxa1<-subset(top_genera_melt3, Genus2 == taxanames[i])
   # taxa1<-subset(top_genera_melt3, Genus2 == "Rare taxa")
    
    metadata<-taxa1
    #str(metadata)
    
    #scale variables
    metadata$Seq_depth_nospikein<-as.numeric(scale(metadata$Seq_depth_nospikein))
    
    metadata$TotalAbundance<-log10(metadata$TotalAbundance)

    
    str(metadata)
    
    # fit gamm    
    m_taxa <- mgcv::bam(log10(Abundance2+1)~
        s(HoursAfterSunrise, bs = "cr")+
        s(month, bs = "cc")+
        s(AgeAtSamplingYears, bs = "cr")+
        #s(Seq_depth_nospikein, bs = "cr")+
        s(Year, bs = "cr", k = 5)+
         s(TotalAbundance, bs = "cr")+
        #TB_stage+
        #  Condition_weight+
        # Temp_max+
        Storage+
        Seq_run+
        s(IndividID, bs = "re")+
        s(GroupAtSampling, bs = "re"),
      #correlation = corARMA(form = ~ 1|Year, p = 3),
      data=metadata,
      family = gaussian)
    
   print(summary(m_taxa))
   gratia::draw(m_taxa)
   

    confint_df_year<-confint(m_taxa, parm = "s(Year)",  type = "confidence")
    
    
    
    confint_df_year$Taxa<-taxanames[i]
    
    confint_df<-confint_df_year
    
    # convert into real estimate by adding the intercapt (= mean abundance)
    
    confint_df$est_real<-confint_df$est+summary(m_taxa)$p.coeff[1] 
    confint_df$lower_real<- confint_df$lower+summary(m_taxa)$p.coeff[1] 
    confint_df$upper_real<- confint_df$upper+summary(m_taxa)$p.coeff[1] 
    
    ## add effect size and p value
    
       summary<-summary(m_taxa)
      confint_df$effect_size<-  summary$s.table[5,3]
      confint_df$P_val<-  summary$s.table[5,4]

    confint_df
    year_estimates[[i]]<-confint_df
    
    
    
  }, error=function(e){})
}

names(year_estimates)<-taxanames

year_estimates_df<-do.call(rbind, year_estimates)

head(year_estimates_df)

year_estimates_df$Significant <-year_estimates_df$P_val<0.05

year_estimates_df <- year_estimates_df  %>% arrange(-effect_size)
head(year_estimates_df)
year_estimates_unique<-distinct(year_estimates_df, Taxa, .keep_all = T)
dim(year_estimates_unique)
taxa_order<-year_estimates_unique$Taxa
year_estimates_df$Taxa<-factor(year_estimates_df$Taxa, levels = taxa_order)

########## calculate effect size 

year_estimates_df$Year2<-substr(year_estimates_df$Year, 1, 4)

calculate_effect_size<-subset(year_estimates_df, Year2 <2001 | Year2 > 2014)

calculate_effect_size<-calculate_effect_size[,c(3,9,10, 16)]
calculate_effect_size$Period <- ifelse(calculate_effect_size$Year2 <2001, "EARLY", "LATE")
table(calculate_effect_size$Period)

calculate_effect_size<-calculate_effect_size %>% group_by(Taxa, Period) %>% summarize(Mean_est = mean(est_real))

calculate_effect_size_wide<- spread(calculate_effect_size, Period, Mean_est)
head(calculate_effect_size_wide)


calculate_effect_size_wide$Effect<- calculate_effect_size_wide$LATE - calculate_effect_size_wide$EARLY

calculate_effect_size_wide<-merge(calculate_effect_size_wide, year_estimates_df, by = "Taxa")

calculate_effect_size_wide<-distinct(calculate_effect_size_wide, Taxa, .keep_all = T)

ggplot(calculate_effect_size_wide, aes(y = reorder(Taxa, Effect), x = Effect))+geom_bar(stat = "identity", aes(fill = Significant))

head(year_estimates_unique)


calculate_effect_size_wide<-calculate_effect_size_wide %>%arrange(Effect)

taxa_order<-calculate_effect_size_wide$Taxa


################### plot 
################### plot 
################### plot 

# first add class column so can match colours with earlier figs

head(calculate_effect_size_wide)

taxonomy<-data.frame(tax_table(genus_ps))
tax<-taxonomy[,c("Genus", "Class")]
tax<-distinct(tax, Genus, .keep_all = T)
head(tax)

calculate_effect_size_wide$Class <- vlookup(calculate_effect_size_wide$Taxa, tax, lookup_column = "Genus", result_column = "Class")

head(calculate_effect_size_wide)
calculate_effect_size_wide$Class_plot<-ifelse(calculate_effect_size_wide$Class %in% top_class, calculate_effect_size_wide$Class, "Other")

calculate_effect_size_wide$Class_plot<-factor(calculate_effect_size_wide$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli") )

## 
calculate_effect_size_wide$Sig <-ifelse(calculate_effect_size_wide$Significant==T, "yes", "no")


plot_tb_1<-ggplot(subset(calculate_effect_size_wide, Taxa != "uncultured"), aes(y = reorder(Taxa, Effect), x = Effect))+
  theme_bw(base_size = 14)+
  geom_bar_pattern(stat = "identity", aes(fill = Class_plot, pattern = Sig),
                   color = "black", 
                   pattern_fill = "black",
                   pattern_angle = 45,
                   pattern_density = 0.1,
                   pattern_spacing = 0.025,
                   pattern_key_scale_factor = 0.6)+
  
   scale_pattern_manual(values = c(no = "stripe", yes = "none"))+
  scale_fill_manual(values = rev(pal))+

  ylab("") +
  theme(plot.margin=unit(c(1,0.2,0.2,0.2),"cm"))+
  xlab("Effect size")+
  labs(pattern = "Significant")+
  labs(fill = "Bacterial class")




```

#### TB/condition/temp effects when year is a random effect

```{r}

TB_estimates<-list()
Condition_estimates<-list()
Temp_estimates<-list()
Rainfall_estimates<-list()

taxanames<-unique(top_genera_melt3$Genus2)

for (i in 1:length(taxanames)){
  tryCatch({ #catch errors
    print(i)
  
    taxa1<-subset(top_genera_melt3, Genus2 == taxanames[i])
   # taxa1<-subset(top_genera_melt3, Genus2 == "Fusobacterium")
    
    metadata<-taxa1
    #str(metadata)
    
    #scale variables
    metadata$Seq_depth_nospikein<-as.numeric(scale(metadata$Seq_depth_nospikein))
    metadata$TB_stage<-factor(metadata$TB_stage, levels = c("Unexposed", "Exposed","Diseased"))
    
    #str(metadata)
    metadata$Year<-factor(metadata$Year)
    metadata$TotalAbundance<-log10(metadata$TotalAbundance)
    metadata$Rainfall<-log10(metadata$sum_rainfall_2weeks+1)
    
    # fit gamm    
    m_taxa <- mgcv::bam(log10(Abundance2+1)~
        s(HoursAfterSunrise, bs = "cr")+
        s(month, bs = "cc")+
        s(AgeAtSamplingYears, bs = "cr")+
       # s(Seq_depth_nospikein, bs = "cr")+
        s(TotalAbundance, bs = "cr")+
        TB_stage+
        Condition_weight+
        Temp_max+
        Rainfall+
        Season+
        Storage+
        Seq_run+
        s(IndividID, bs = "re")+
        s(Year, bs = "re")+
        s(GroupAtSampling, bs = "re"),
      #correlation = corARMA(form = ~ 1|Year, p = 3),
      data=metadata, 
      family = gaussian)
    
   print(summary(m_taxa))
   
       summary<-summary(m_taxa)
     
    ## TB estimates 
    
    TB_estimates_df<-data.frame(summary$p.table)[1:3,]
     TB_estimates_df$Level<-c("Unexposed", "Exposed", "Diseased")
     names(TB_estimates_df)[4]<-"P_val"
     
    TB_estimates_df$Estimate2<-ifelse(TB_estimates_df$Level == "Unexposed", 0, TB_estimates_df$Estimate)
         
     
      TB_estimates_df$lower<-NA
      TB_estimates_df$upper<-NA
    
      # confidence intervals exposed
      beta <- coef(m_taxa)
      Vb <- vcov(m_taxa, unconditional = TRUE)
     se <- sqrt(diag(Vb))
      j <- which(names(beta) == "(Intercept)")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[1,7]<-cis[1]
      TB_estimates_df[1,8]<-cis[2]
      
            # confidence intervals exposed
      j <- which(names(beta) == "TB_stageExposed")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[2,7]<-cis[1]
      TB_estimates_df[2,8]<-cis[2]
     
                 # confidence intervals symptomatic
      j <- which(names(beta) == "TB_stageDiseased")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[3,7]<-cis[1]
      TB_estimates_df[3,8]<-cis[2]
      
      ## add coefficient
      
      TB_estimates_df$Coef <-as.numeric(coef(m_taxa)[1:3])
        TB_estimates_df$Taxa <-taxanames[i]
    TB_estimates[[i]]<-TB_estimates_df
      
      ## extract condition stats
      
      condition_estimates<- data.frame(summary$p.table)[4,]
       condition_estimates$Taxa<-taxanames[i]
       Condition_estimates[[i]]<-condition_estimates
      
          ## extract max temp stats
      
      temp_estimates<- data.frame(summary$p.table)[5,]
       temp_estimates$Taxa<-taxanames[i]
       Temp_estimates[[i]]<-temp_estimates
       
       ## extract rainfall stats
      
      rainfall_estimates<- data.frame(summary$p.table)[6,]
       rainfall_estimates$Taxa<-taxanames[i]
       Rainfall_estimates[[i]]<-rainfall_estimates
  
    
  }, error=function(e){})
}

```

### TB

```{r}
##########
##########
##########
##########

names(TB_estimates)<-taxanames

TB_estimates_df<-do.call(rbind, TB_estimates)

head(TB_estimates_df)

TB_estimates_df$Significant <-TB_estimates_df$P_val < 0.05

# edit significance (intercept will always be significant)

sig_taxa<-subset(TB_estimates_df, Significant == T & Level !="Unexposed")
sig_taxa<-unique(sig_taxa$Taxa)

TB_estimates_df$Significant<- TB_estimates_df$Taxa %in%  sig_taxa


TB_estimates_df

## edit confidence intervals

TB_estimates_df$Level<-factor(TB_estimates_df$Level, levels = c("Diseased", "Exposed", "Unexposed"))

TB_estimates_df$lower<- ifelse(TB_estimates_df$Level == "Unexposed", TB_estimates_df$lower-TB_estimates_df$Estimate, TB_estimates_df$lower)

TB_estimates_df$upper<- ifelse(TB_estimates_df$Level == "Unexposed", TB_estimates_df$upper-TB_estimates_df$Estimate, TB_estimates_df$upper)


# order taxa

TB_estimates_df$Taxa<-factor(TB_estimates_df$Taxa, levels = taxa_order)



######
#############
#############
#############

plot_tb_2<-ggplot(subset(TB_estimates_df, Taxa!= "uncultured"), aes( x = Estimate2, y = Taxa, group = Level))+
  theme_bw(base_size = 14)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  geom_errorbarh(aes(col = Level, xmin = Estimate2- Std..Error, xmax = Estimate2+ Std..Error), height = 0, size = 1, position=position_dodge(width=0.5))+
  
    geom_errorbarh(aes(alpha = Significant, xmin = Estimate2- Std..Error, xmax = Estimate2+ Std..Error), height = 0, size = 1, col = "grey", position=position_dodge(width=0.5))+
  
  scale_alpha_discrete(range = c(1,0), guide = "none")+

   geom_point(aes(col = Level), size = 2, position=position_dodge(width=0.5))+
  
  scale_color_manual(values = c( "brown2", "skyblue","forestgreen"),  guide = guide_legend(reverse = TRUE))+
  xlim(c(-0.9,NA))+

  theme(axis.text.y = element_blank(), axis.title.y = element_blank())+
  xlab("Effect size")+
  labs(col = "TB stage")+
  theme(plot.margin=unit(c(1,0.2,0.2,0.2),"cm"))+
  # legend
  theme( legend.background = element_blank(),
        legend.box.background = element_rect(colour = "grey", fill = "white"))+ 
  theme(legend.position=c(0.2,0.95),  legend.direction="vertical", 
       legend.key.height = unit(0.05, 'cm'),
        legend.key.width = unit(0.02, 'cm'),
        legend.title=element_text(size=8),
       legend.text =element_text(size=8))+
   guides(colour=guide_legend(title.position="top",                           title.hjust =0.5,reverse = TRUE))+
  coord_cartesian(xlim = c(-0.7, 0.7))  +
  scale_x_continuous(breaks = c(-0.3, 0, 0.3))+
 theme(legend.margin=margin(t = c(0.1,0.1,0.1,0.1), unit='cm'))

```

### Body condition

```{r}
names(Condition_estimates)<-taxanames

condition_estimates_df<-do.call(rbind, Condition_estimates)

condition_estimates_df$Significant <-condition_estimates_df$Pr...t.. < 0.05


condition_estimates_df$Taxa<-factor(condition_estimates_df$Taxa, levels = taxa_order)

ggplot(subset(condition_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  geom_point( aes(col = Significant), size = 3)+
  theme_bw()+
  geom_errorbarh(aes(col = Significant, xmin = Estimate- Std..Error, xmax = Estimate+ Std..Error), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "red"))
  
plot_condition<-
ggplot(subset(condition_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  theme_bw(base_size = 14)+
  geom_point( aes(col = Significant), size = 3)+
  geom_errorbarh(aes(col = Significant, xmin = ((Estimate- Std..Error)-0.0005), xmax = ((Estimate+ Std..Error)+0.0005) ), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "blue"))+
   theme(axis.text.y = element_blank(), axis.title.y = element_blank())+
  xlab("Effect size")+
  xlim(c(-0.004,NA))+
  #legend
  theme( legend.background = element_blank(),
        legend.box.background = element_rect(colour = "grey", fill = "white"))+
theme(legend.position=c(0.8,0.95), legend.direction="vertical", 
      
       legend.key.height = unit(0.02, 'cm'),
        legend.key.width = unit(0.02, 'cm'),
        legend.title=element_text(size=8),
       legend.text =element_text(size=6))+
   guides(colour=guide_legend(title.position="top", reverse = T,
                                     title.hjust =0.5))+
  coord_cartesian(xlim = c(-0.01, 0.01))+
  scale_x_continuous(breaks = c(-0.005, 0, 0.005))

```

### Maximum temp

```{r}
names(Temp_estimates)<-taxanames

temp_estimates_df<-do.call(rbind, Temp_estimates)

temp_estimates_df$Significant <-temp_estimates_df$Pr...t.. < 0.05


temp_estimates_df$Taxa<-factor(temp_estimates_df$Taxa, levels = taxa_order)

ggplot(subset(temp_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  geom_point( aes(col = Significant), size = 3)+
  theme_bw()+
  geom_errorbarh(aes(col = Significant, xmin = Estimate- Std..Error, xmax = Estimate+ Std..Error), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "red"))


  
plot_temp<-
ggplot(subset(temp_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  theme_bw(base_size = 14)+
  geom_point( aes(col = Significant), size = 3)+
  geom_errorbarh(aes(col = Significant, xmin = Estimate- Std..Error, xmax = Estimate+ Std..Error), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "blue"))+
   theme(axis.text.y = element_blank(), axis.title.y = element_blank())+
  xlab("Effect size")+
 # xlim(c(-0.004,NA))+
  #legend
   theme( legend.background = element_blank(),
        legend.box.background = element_rect(colour = "grey", fill = "white"))+
   theme(legend.position=c(0.8,0.95), legend.direction="vertical", 
       legend.key.height = unit(0.02, 'cm'),
        legend.key.width = unit(0.02, 'cm'),
        legend.title=element_text(size=8),
       legend.text =element_text(size=6))+
   guides(colour=guide_legend(title.position="top", reverse = T,
                                     title.hjust =0.5))+
  coord_cartesian(xlim = c(-0.1, 0.1))+
  scale_x_continuous(breaks = c(-0.05, 0, 0.05))

```

### Rainfall

```{r}
names(Rainfall_estimates)<-taxanames

rainfall_estimates_df<-do.call(rbind, Rainfall_estimates)

rainfall_estimates_df$Significant <-rainfall_estimates_df$Pr...t.. < 0.05


rainfall_estimates_df$Taxa<-factor(rainfall_estimates_df$Taxa, levels = taxa_order)

ggplot(subset(rainfall_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  geom_point( aes(col = Significant), size = 3)+
  theme_bw()+
  geom_errorbarh(aes(col = Significant, xmin = Estimate- Std..Error, xmax = Estimate+ Std..Error), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "red"))


  
plot_rainfall<-
ggplot(subset(rainfall_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  theme_bw(base_size = 14)+
  geom_point( aes(col = Significant), size = 3)+
  geom_errorbarh(aes(col = Significant, xmin = Estimate- Std..Error, xmax = Estimate+ Std..Error), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "blue"))+
   theme(axis.text.y = element_blank(), axis.title.y = element_blank())+
  xlab("Effect size")+
 # xlim(c(-0.004,NA))+
  #legend
   theme( legend.background = element_blank(),
        legend.box.background = element_rect(colour = "grey", fill = "white"))+
    theme(legend.position=c(0.8,0.95), legend.direction="vertical", 
       legend.key.height = unit(0.02, 'cm'),
        legend.key.width = unit(0.02, 'cm'),
        legend.title=element_text(size=8),
       legend.text =element_text(size=6))+
   guides(colour=guide_legend(title.position="top", reverse = T,
                                     title.hjust =0.5))+
  coord_cartesian(xlim = c(-0.5, 0.5))+
  scale_x_continuous(breaks = c(-0.3, 0, 0.3))

```



### Plot together

```{r}
plot_tb_1
plot_tb_2
plot_condition
plot_temp
plot_rainfall

# Merging legends
legend_1 <- get_legend(plot_tb_1)
#legend_2 <- get_legend(plot_)

#legends <- ggarrange(legend_1, legend_2, nrow=2)

# Combining plots
rm_legend <- function(p){p + theme(legend.position = "none")}

plots <- ggarrange(rm_legend(plot_tb_1), plot_tb_2, plot_rainfall, plot_temp, plot_condition, nrow = 1, align = "h", widths = c(2.2,1,1,1, 1), labels = c("a) Year", "b) TB", "c) Rainfall","d) Max. temp.", "e) Body condition"), label.x = c(0.45,-0.05,-0.15,-0.25, -0.32))

# figure 3
ggarrange(plots, legend_1, widths = c(0.85, 0.15))

ggsave("Figures/Fig4.pdf")



```


# Network analysis

Need to do CLR analysis from here as this accounts for bacterial load

```{r}

genus_core<-prune_taxa(top_genera, genus_ps)



#### extract site x species array

phylo<-genus_core

taxtable<-data.frame(tax_table(phylo))

taxa_names(phylo)<-taxtable$Genus

phylo_array<-data.frame(t(phylo@otu_table@.Data))


#phylo_array[1:5,1:5]
names(phylo_array)<-taxa_names(phylo)

dim(phylo_array)

########################## 

#### genreate association matrix from the 'real' data
# this uses the package NetCoMi


net_real <- netConstruct(phylo_array, verbose = 0, 
  measure = "spearman",
  normMethod = "clr",
  zeroMethod = "none",
  sparsMethod = "threshold",
  thresh = 0.3)

str(net_real)



props_single <- netAnalyze(net_real, 
                           centrLCC = TRUE,
                           clustMethod = "cluster_fast_greedy",
                           hubPar = "eigenvector",
                           weightDeg = FALSE, 
                           normDeg = FALSE)



plot(props_single, 
          nodeColor = "cluster", 
     layout = "spring",
     repulsion = 1,
         # nodeSize = "eigenvector",
          title1 = F, 
          #showTitle = TRUE,
          cexTitle = 2.3,
     shortenLabels = "simple",
     labelScale = F,
     nodeSize = "fix",
     colorVec = c("orchid", "skyblue", "orange", "lightgreen", "pink"),
     nodeTransp = 0,
     borderCol = "white",
     borderWidth = 3,
    highlightHubs = F ,
    edgeTranspLow = 0)

legend(0.7, 1.1, cex = 2, title = "Direction", 
       legend = c("+","-"), lty = 1, lwd = 2, col = c("#009900","red"), 
       bty = "n", horiz = F)

cluster_memb<-data.frame(props_single$clustering$clust1)

head(cluster_memb, 20)
names(cluster_memb)<-"Cluster"

# make singletons their own cluster (not zero)

cluster_memb$Cluster<-ifelse(row.names(cluster_memb)=="Escherichia-Shigella", 6,cluster_memb$Cluster )
cluster_memb$Cluster<-ifelse(row.names(cluster_memb)=="Peptoclostridium", 7,cluster_memb$Cluster )
cluster_memb$Cluster<-ifelse(row.names(cluster_memb)=="Ruminococcaceae UCG-014", 8,cluster_memb$Cluster )


cluster_memb$Cluster2<-case_when(cluster_memb == 1 ~ "Lactococcus cluster",
                                 cluster_memb == 2 ~ "Bacteroides cluster",
                                 cluster_memb == 3 ~ "Blautia cluster",
                                 cluster_memb == 4 ~ "Clostridium cluster",
                                 cluster_memb == 5 ~ "Bacillus cluster",
                                 cluster_memb == 6 ~ "Escherichia-Shigella",
                                 cluster_memb == 7 ~ "Peptoclostridium",
                                 cluster_memb == 8 ~ "Ruminococcus UCG-014" )

```

# Survival analysis

```{r}


# add death dates because they are missing

############################################

meerkat_filtered_genus<-tax_glom(meerkat_filtered, taxrank = "Genus")
genus_core<-prune_taxa(top_genera, meerkat_filtered_genus)
taxtable<-data.frame(tax_table(genus_core))
taxa_names(genus_core)<-taxtable$Genus


# 1) Genus level

survival_df<-data.frame(sample_data(genus_core))

survival_df<-survival_df[,c("IndividID", "feature.id", "SampleNo", "Storage", "Seq_run", "SampleDate","Year", "HoursAfterSunrise", "Sex", "BirthDate", "DeathDate", "AgeAtSampling", "DaysTilDeath", "AgeAtDeath", "GroupAtSampling", "Weight", "Condition_weight", "SocialStatus", "Season","sum_rainfall_month", "Temp_max", "TB_status", "TB_exposure", "TB_resistance", "TB_survival", "TB_stage", "TotalAbundance")]

# calculate if ID is alive 6 months later, TRUE = dead

survival_df$Time_180<-survival_df$SampleDate+180
survival_df$Survival_180<- ifelse(survival_df$Time_180 < survival_df$DeathDate, FALSE,TRUE)


table(survival_df$Survival_180)
survival_df$FollowUp_180<-180


## add on genus level abundance 
genus_clr<-microbiome::transform(genus_core, transform = "clr")

otutable<-data.frame(t(otu_table(genus_clr)))
otutable[1:5, 1:5]
names(otutable)<-taxa_names(genus_clr)
names(otutable)<-make.names(names(otutable),unique = TRUE)
otutable<-data.frame(scale(otutable))
#hist(otutable$Helicobacter)

survival_df_genus<-merge(survival_df, otutable, by = 0)
head(survival_df_genus)

# model 

names(survival_df_genus)[32:60]
names(survival_df_genus)

survival_df_genus<-survival_df_genus[,c(2, 30, 31,8, 13,20,27,19,18,32:60)]
names(survival_df_genus)
survival_df_genus$AgeAtSampling<-scale(log10(survival_df_genus$AgeAtSampling))
survival_df_genus$Condition_weight<-scale(survival_df_genus$Condition_weight)
survival_df_genus$IndividID<-as.character(survival_df_genus$IndividID)
survival_df_genus$Season <-case_when(survival_df_genus$Season == "WET" ~ "Wet",
                                     survival_df_genus$Season == "DRY" ~ "Dry")
names(survival_df_genus)[5]<-"Age"

survival_df_genus<-subset(survival_df_genus, !is.na(Survival_180))
dim(survival_df_genus)

#survival_df_genus$Survival<-Surv(survival_df_genus$FollowUp_180, survival_df_genus$Survival_180)


frmla <- as.formula(paste("Surv(FollowUp_180, Survival_180)", 
                              paste(names(survival_df_genus)[5:38],  sep = "", 
                                    collapse = " + "), sep = " ~ "))
    
    frmla <-update.formula(frmla, ~ . + (1|IndividID))

    

fit_genus <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + Season + TB_stage + 
    SocialStatus + Condition_weight + 
    Fusobacterium+
      Bacteroides+
      Parabacteroides+
      Phascolarctobacterium+
        Helicobacter + 
      Alloprevotella+

       Enterococcus + 
      Pediococcus + 
      Lactococcus+
      Peptoclostridium+
      Collinsella+
      Slackia+
      
      (1 | IndividID),
              data = survival_df_genus)


fit_genus

sjPlot::plot_model(fit_genus, axis.lim = c(0.8,20), vline.color = "darkgrey", sort.est = T)+  theme_bw(base_size = 16)+ggtitle("")+ylab("Hazard ratio")

options(na.action = "na.fail") 

dredge_genus_df<- MuMIn::dredge(fit_genus, fixed = c("Age", "Season", "TB_stage",
    "SocialStatus", "Condition_weight"))

#write.csv(dredge_genus_df, "dredge_results_genus.csv")


fit_genus_early <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + Season + TB_stage + 
    SocialStatus + Condition_weight + Helicobacter + Geodermatophilus + 
    Fusobacterium + Tyzzerella.4 + X.Ruminococcus..torques.group + 
    Blautia + Marvinbryantia + 
      Lachnoclostridium + Lachnospiraceae.NK4A136.group + 
    Rikenellaceae.RC9.gut.group + Bacteroides + Alloprevotella + 
    Parabacteroides + Paeniclostridium + Romboutsia + Peptoclostridium + 
    Collinsella + Enterococcus + Pediococcus + Weissella + Bacillus + 
    Turicibacter + Catenisphaera + Lactococcus + Clostridium.sensu.stricto.1 + 
    Peptococcus + Phascolarctobacterium + Ruminococcaceae.UCG.014 + 
    Slackia + (1 | IndividID),
              data = survival_df_genus, 
  subset=(Year<2009))

sjPlot::plot_model(fit_genus_early, axis.lim = c(0.8,20), vline.color = "darkgrey", sort.est = T)+  theme_bw(base_size = 16)+ggtitle("")

fit_genus_late <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + Season + TB_stage + 
    SocialStatus + Condition_weight + Helicobacter + Geodermatophilus + 
    Fusobacterium + Tyzzerella.4 + X.Ruminococcus..torques.group + 
    Blautia + Marvinbryantia + 
      Lachnoclostridium + Lachnospiraceae.NK4A136.group + 
    Rikenellaceae.RC9.gut.group + Bacteroides + Alloprevotella + 
    Parabacteroides + Paeniclostridium + Romboutsia + Peptoclostridium + 
    Collinsella + Enterococcus + Pediococcus + Weissella + Bacillus + 
    Turicibacter + Catenisphaera + Lactococcus + Clostridium.sensu.stricto.1 + 
    Peptococcus + Phascolarctobacterium + Ruminococcaceae.UCG.014 + 
    Slackia + (1 | IndividID),
              data = survival_df_genus, 
  subset=(Year>2008))

sjPlot::plot_model(fit_genus_late, axis.lim = c(0.5,20), vline.color = "darkgrey", sort.est = T)+  theme_bw(base_size = 16)+ggtitle("")




######### 3) Network ############
######### 3) Network ############
######### 3) Network ############
######### 3) Network ############

pseq_network<-genus_core
head(cluster_memb)
tail(cluster_memb)
pseq_network@tax_table@.Data[,1]<-cluster_memb$Cluster2

pseq_network@tax_table@.Data

pseq_network2<-tax_glom(pseq_network, taxrank = "Kingdom")
pseq_network2<-microbiome::transform(pseq_network2, "clr")
taxa_names(pseq_network2)
taxa_names(pseq_network2)<-c("Blautia_cluster", "Bacteroides_cluster", "Peptoclostridium", "Lactococcus_cluster", "Bacillus_cluster", "Clostridium_cluster", "Ruminococcus_UCG-014", "Escherichia-Shigella")


otutable<-data.frame(t(otu_table(pseq_network2)))
otutable[1:5, 1:5]

otutable<-data.frame(scale(otutable))
#hist(otutable$Helicobacter)

survival_df_network<-merge(survival_df, otutable, by = 0)



#############
#############
#############
#############

names(survival_df_network)

survival_df_network<-survival_df_network[,c(2, 8,30, 31, 13,20,27,19,18,32:38)]
names(survival_df_network)
survival_df_network$AgeAtSampling<-scale(log10(survival_df_network$AgeAtSampling))
survival_df_network$Condition_weight<-scale(survival_df_network$Condition_weight)
survival_df_network$IndividID<-as.character(survival_df_network$IndividID)

survival_df_network$Season <-case_when(survival_df_network$Season == "WET" ~ "Wet",
                                     survival_df_network$Season == "DRY" ~ "Dry")
names(survival_df_network)[5]<-"Age"

survival_df_network<-subset(survival_df_network, !is.na(Survival_180))
dim(survival_df_network)


#survival_df_network$Survival<-Surv(survival_df_network$FollowUp_180, survival_df_network$Survival_180)


frmla <- as.formula(paste("Surv(FollowUp_180, Survival_180)", 
                              paste(names(survival_df_network)[5:16],  sep = "", 
                                    collapse = " + "), sep = " ~ "))
    
    frmla <-update.formula(frmla, ~ . + (1|IndividID))


fit_network <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + 
                              Season + 
                              TB_stage + 
    SocialStatus + 
    Condition_weight + 
      Blautia_cluster + 
      Bacteroides_cluster + 
    Peptoclostridium + 
      Lactococcus_cluster + Bacillus_cluster + 
    Clostridium_cluster + Ruminococcus_UCG.014  + 
    (1 | IndividID),
     data = survival_df_network)

summary(fit_network)


sjPlot::plot_model(fit_network, axis.lim = c(0.8,11), 
                   vline.color = "darkgrey", sort.est = T, 
                   dot.size= 4, line.size = 1.5, show.p = T)+  
  theme_bw(base_size = 20)+ggtitle("")+ylab("Hazard ratio")


plot(props_single, 
          nodeColor = "cluster", 
     layout = "layout_with_fr",
         # nodeSize = "eigenvector",
          title1 = F, 
          #showTitle = TRUE,
          cexTitle = 2.3,
     shortenLabels = "none",
     labelScale = F,
     nodeSize = "degree")

dredge_network_df<- MuMIn::dredge(fit_network, fixed = c("Age", "Season", "TB_stage",
    "SocialStatus", "Condition_weight"))

#write.csv(dredge_network_df, "dredge_results_network.csv")


## early vs late

fit_network_early <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + 
                              Season + 
                              TB_stage + 
    SocialStatus + 
    Condition_weight + 
      Blautia_cluster + 
      Bacteroides_cluster + 
    Peptoclostridium + 
      Lactococcus_cluster + Bacillus_cluster + 
    Clostridium_cluster + Ruminococcus_UCG.014  + 
    (1 | IndividID),
              data = survival_df_network, 
  subset=(Year<2009))

sjPlot::plot_model(fit_network_early, axis.lim = c(0.8,20), vline.color = "darkgrey", sort.est = T)+  theme_bw(base_size = 16)+ggtitle("")

fit_network_late <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + 
                              Season + 
                              TB_stage + 
    SocialStatus + 
    Condition_weight + 
      Blautia_cluster + 
      Bacteroides_cluster + 
    Peptoclostridium + 
      Lactococcus_cluster + Bacillus_cluster + 
    Clostridium_cluster + Ruminococcus_UCG.014  + 
    (1 | IndividID),
              data = survival_df_network, 
  subset=(Year>2008))

sjPlot::plot_model(fit_network_late, axis.lim = c(0.5,20), vline.color = "darkgrey", sort.est = T)+  theme_bw(base_size = 16)+ggtitle("")



```


########################

```{r}


####### before and after 2008

fit_early <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + Season + TB_stage + 
    SocialStatus + Condition_weight + Blautia_cluster + Bacteroides_cluster + 
    Peptoclostridium + Lactococcus_cluster + Bacillus_cluster + 
    Clostridium_cluster + Ruminococcus_UCG.014  + 
    (1 | IndividID),
     data = survival_df_network, 
              subset=(Year<2009))

summary(fit_early)


P_early<-sjPlot::plot_model(fit_early, axis.lim = c(0.8,20), vline.color = "darkgrey", sort.est = T)+
  theme_bw(base_size = 14)+ggtitle("")
P_early



### late

fit_late <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ Age + Season + TB_stage + 
    SocialStatus + Condition_weight + Blautia_cluster + Bacteroides_cluster + 
    Peptoclostridium + Lactococcus_cluster + Bacillus_cluster + 
    Clostridium_cluster + Ruminococcus_UCG.014  + 
    (1 | IndividID),
     data = survival_df_network, 
              subset=(Year>2008))
summary(fit_late)


dim(subset(survival_df, !is.na(Survival_180) & Year<2009))
dim(subset(survival_df, !is.na(Survival_180) & Year>2008))

P_late<-sjPlot::plot_model(fit_late, axis.lim = c(0.5,20), vline.color = "darkgrey", sort.est = T)+
  theme_bw(base_size = 14)+ggtitle("")

P_late


ggarrange(P_early, P_late, ncol = 2, labels = c("a) Before 2008", "b) After 2008"))


################ beta diversity ##########
################ beta diversity ##########
################ beta diversity ##########
################ beta diversity ##########

fit_beta1 <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ 
              scale(WU1)+
              scale(WU2)+
              scale(WU3)+
              scale(WU4)+
                 
                  scale(log10(Age))+
                 SocialStatus+
                 TB_stage+
                 Season+
                (1|IndividID), 
              survival_df, 
              subset=(!is.na(Survival_180)))

sjPlot::plot_model(fit_beta1, axis.lim = c(0.8,20), vline.color = "darkgrey", sort.est = T)+
  theme_bw(base_size = 14)+ggtitle("")
names(survival_df)


fit_beta2 <- coxme::coxme(Surv(FollowUp_180, Survival_180) ~ 
              scale(UU1)+
              scale(UU2)+
              scale(UU3)+
              scale(UU4)+
                 
                  scale(log10(Age))+
                 SocialStatus+
                 TB_stage+
                 Season+
                (1|IndividID), 
              survival_df, 
              subset=(!is.na(Survival_180)))

sjPlot::plot_model(fit_beta2, axis.lim = c(0.8,20), vline.color = "darkgrey", sort.est = T)+
  theme_bw(base_size = 14)+ggtitle("")

summary(fit_beta1)
summary(fit_beta2)

```


# 9. CLR transformation

```{r}
meerkat_final<-readRDS("C:\\Users\\risel\\Dropbox\\Sommer postdoc\\Meerkat project\\PROJECTS\\MeerkatSpikeInData\\Analysis updated\\DATA 2021\\meerkat_unrarefied.rds")


meerkat_genus<-tax_glom(meerkat_final, taxrank = "Genus")
#meerkat_genus<-prune_taxa(top_genera, meerkat_genus)

sum(taxa_sums(meerkat_genus))
sum(taxa_sums(meerkat_final))

meerkat_clr <- microbiome::transform(meerkat_genus, "clr")
taxa_names(meerkat_clr)<-data.frame(tax_table(meerkat_clr))$Genus

sample_data(meerkat_clr)<-sample_data(meerkat_clr)[,1:56]



### change some variables 

sample_data(meerkat_clr)$TimeCat<-ifelse(sample_data(meerkat_clr)$HoursAfterSunrise < 7, "MORNING", "AFTERNOON")
table(sample_data(meerkat_clr)$TimeCat)
meerkat_clr

head(sample_data(meerkat_clr))



# generate categorical variable for wet and dry seasons (May-Sep = dry, Oct-April = Wet)

sample_data(meerkat_clr)$month <-lubridate::month(sample_data(meerkat_clr)$SampleDate)

sample_data(meerkat_clr)$Season <- as.factor(ifelse(sample_data(meerkat_clr)$month == 5| 
    sample_data(meerkat_clr)$month == 6 |
    sample_data(meerkat_clr)$month == 7 |
    sample_data(meerkat_clr)$month == 8 |
    sample_data(meerkat_clr)$month == 9  , "DRY", "WET"))

## TB status

levels(sample_data(meerkat_clr)$TB_status)
levels(sample_data(meerkat_clr)$TB_stage)

sample_data(meerkat_clr)$TB_status<- factor(sample_data(meerkat_clr)$TB_status, levels = c("unexposed", "exposed_asymptomatic", "exposed_symptomatic"))

levels(sample_data(meerkat_clr)$TB_status)<-c("Unexposed", "Exposed", "Diseased")


### TB stage

sample_data(meerkat_clr)$TB_stage<- factor(sample_data(meerkat_clr)$TB_stage, levels = c("TB_uninfected", "TB_exposed", "TB_symptomatic"))

levels(sample_data(meerkat_clr)$TB_stage)<-c("Unexposed", "Exposed", "Diseased")




```

```{r}

genus_abundances_df<-data.frame(taxa_sums(meerkat_clr))

genus_abundances_df$ASV<-row.names(genus_abundances_df)

taxonomy<-data.frame(tax_table(meerkat_clr))
head(taxonomy)
taxonomy$ASV<-row.names(taxonomy)


genus_abundances_df<-merge(genus_abundances_df, taxonomy, by = "ASV")
names(genus_abundances_df)[2]<-"Abundance"

genus_abundances_df<-genus_abundances_df %>% arrange(-Abundance)

top_genera<- genus_abundances_df$ASV[1:30]

top_meerkat_clr<-prune_taxa(top_genera, meerkat_clr)

sample_data(top_meerkat_clr)<-sample_data(top_meerkat_clr)[,c( "feature.id")]

top_genera_melt<-psmelt(top_meerkat_clr)
dim(top_genera_melt)

metadata<-data.frame(sample_data(meerkat_clr))
#metadata<-data.frame(sample_data(subset_samples(meerkat_clr, Storage == "FREEZEDRIED")))

top_genera_melt2<- merge(top_genera_melt, metadata, by = "feature.id")
names(top_genera_melt2)
head(top_genera_melt2)
unique(top_genera_melt2$Genus)



```

-   Year estimates

```{r}

# loop to fit GAMMs to each genus

year_estimates<-list()
TB_estimates<-list()

taxanames<-unique(top_genera_melt2$Genus)

for (i in 1:length(taxanames)){
  tryCatch({ #catch errors
    print(i)
  
    taxa1<-subset(top_genera_melt2, Genus == taxanames[i])
   # taxa1<-subset(top_genera_melt2, Genus == "Blautia")
    
    metadata<-taxa1
    #str(metadata)
    
    #scale variables
    metadata$Seq_depth_nospikein<-as.numeric(scale(metadata$Seq_depth_nospikein))
    metadata$TB_stage<-factor(metadata$TB_stage, levels = c("Unexposed", "Exposed","Diseased"))
    
    #str(metadata)
    
    # fit gamm    
    m_taxa <- mgcv::bam(Abundance~
        s(HoursAfterSunrise, bs = "cr")+
        s(month, bs = "cc")+
        s(AgeAtSamplingYears, bs = "cr")+
        s(Seq_depth_nospikein, bs = "cr")+
        s(Year, bs = "cr", k = 5)+
        # s(TotalAbundance, bs = "cr")+
        TB_stage+
       Condition_weight+
        Storage+
         Condition_weight+
        Seq_run+
        s(IndividID, bs = "re")+
        s(GroupAtSampling, bs = "re"),
      #correlation = corARMA(form = ~ 1|Year, p = 3),
      data=metadata,
      family = gaussian)
    
   print(summary(m_taxa))
   

    confint_df_year<-confint(m_taxa, parm = "s(Year)",  type = "confidence")
    
    
    
    confint_df_year$Taxa<-taxanames[i]
    
    confint_df<-confint_df_year
    
    # convert into real estimate by adding the intercapt (= mean abundance)
    
    confint_df$est_real<-confint_df$est+summary(m_taxa)$p.coeff[1] 
    confint_df$lower_real<- confint_df$lower+summary(m_taxa)$p.coeff[1] 
    confint_df$upper_real<- confint_df$upper+summary(m_taxa)$p.coeff[1] 
    
    ## add effect size and p value
    
       summary<-summary(m_taxa)
      confint_df$effect_size<-  summary$s.table[5,3]
      confint_df$P_val<-  summary$s.table[5,4]

    confint_df
    year_estimates[[i]]<-confint_df
    
    
    ## TB estimates 
    
    TB_estimates_df<-data.frame(summary$p.table)[1:3,]
     TB_estimates_df$Level<-c("Unexposed", "Exposed", "Diseased")
     names(TB_estimates_df)[4]<-"P_val"
     
         TB_estimates_df$Estimate2<-ifelse(TB_estimates_df$Level == "Unexposed", 0, TB_estimates_df$Estimate)
         
     
      TB_estimates_df$lower<-NA
      TB_estimates_df$upper<-NA
    
      # confidence intervals exposed
      beta <- coef(m_taxa)
      Vb <- vcov(m_taxa, unconditional = TRUE)
     se <- sqrt(diag(Vb))
      j <- which(names(beta) == "(Intercept)")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[1,7]<-cis[1]
      TB_estimates_df[1,8]<-cis[2]
      
            # confidence intervals exposed
      j <- which(names(beta) == "TB_stageExposed")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[2,7]<-cis[1]
      TB_estimates_df[2,8]<-cis[2]
     
                 # confidence intervals symptomatic
      j <- which(names(beta) == "TB_stageDiseased")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[3,7]<-cis[1]
      TB_estimates_df[3,8]<-cis[2]
      
      ## add coefficient
      
      TB_estimates_df$Coef <-as.numeric(coef(m_taxa)[1:3])
      
      
    TB_estimates_df$Taxa <-taxanames[i]
    TB_estimates[[i]]<-TB_estimates_df
    
  }, error=function(e){})
}

names(year_estimates)<-taxanames

year_estimates_df<-do.call(rbind, year_estimates)

head(year_estimates_df)

year_estimates_df$Significant <-year_estimates_df$P_val<0.05

year_estimates_df <- year_estimates_df  %>% arrange(-effect_size)
head(year_estimates_df)
year_estimates_unique<-distinct(year_estimates_df, Taxa, .keep_all = T)
dim(year_estimates_unique)
taxa_order<-year_estimates_unique$Taxa
year_estimates_df$Taxa<-factor(year_estimates_df$Taxa, levels = taxa_order)

########## calculate effect size 

calculate_effect_size<-subset(year_estimates_df, Year == 1997 | Year == 2019)
calculate_effect_size<-calculate_effect_size[,c(3,9,10)]
calculate_effect_size_wide<- spread(calculate_effect_size, Year, est_real)
head(calculate_effect_size_wide)


calculate_effect_size_wide$Effect<- calculate_effect_size_wide$`2019` - calculate_effect_size_wide$`1997`

calculate_effect_size_wide<-merge(calculate_effect_size_wide, year_estimates_df, by = "Taxa")

calculate_effect_size_wide<-distinct(calculate_effect_size_wide, Taxa, .keep_all = T)

ggplot(calculate_effect_size_wide, aes(y = reorder(Taxa, Effect), x = Effect))+geom_bar(stat = "identity", aes(fill = Significant))

head(year_estimates_unique)


calculate_effect_size_wide<-calculate_effect_size_wide %>%arrange(Effect)

taxa_order<-calculate_effect_size_wide$Taxa

```

-   TB estimates when including year as a smoothed factor

```{r}

names(TB_estimates)<-taxanames

TB_estimates_df<-do.call(rbind, TB_estimates)

head(TB_estimates_df)

TB_estimates_df$Significant <-TB_estimates_df$P_val < 0.05

subset(TB_estimates_df, Significant == T)


TB_estimates_df

TB_estimates_df$Level<-factor(TB_estimates_df$Level, levels = c("Diseased", "Exposed", "Unexposed"))

TB_estimates_df$lower<- ifelse(TB_estimates_df$Level == "Unexposed", TB_estimates_df$lower-TB_estimates_df$Estimate, TB_estimates_df$lower)

TB_estimates_df$upper<- ifelse(TB_estimates_df$Level == "Unexposed", TB_estimates_df$upper-TB_estimates_df$Estimate, TB_estimates_df$upper)


ggplot(TB_estimates_df, aes( x = Estimate2, y = Level))+
  geom_point(aes(col = Level), size = 2)+
  theme_bw()+
  geom_errorbarh(aes(col = Level, xmin = Estimate2- Std..Error, xmax = Estimate2+ Std..Error), height = 0, size = 1)+
  facet_wrap(~Taxa, scale = "free_x")+
    scale_color_manual(values = c("red", "bisque2", "skyblue"))

ggplot(TB_estimates_df, aes( x = Estimate2, y = Taxa, group = Level))+
  geom_point(aes(col = Level), size = 2, position=position_dodge(width=0.5))+
  theme_bw()+
  geom_errorbarh(aes(col = Level, xmin = lower, xmax = upper), height = 0, size = 1, position=position_dodge(width=0.5))+
    scale_color_manual(values = c("red", "bisque2", "skyblue"))

TB_estimates_df$Taxa<-factor(TB_estimates_df$Taxa, levels = taxa_order)

plot_tb_2<-ggplot(subset(TB_estimates_df, Taxa!= "uncultured"), aes( x = Estimate2, y = Taxa, group = Level))+
  geom_point(aes(col = Level), size = 3, position=position_dodge(width=0.5))+
  theme_bw()+
  geom_errorbarh(aes(col = Level, xmin = Estimate2- Std..Error, xmax = Estimate2+ Std..Error), height = 0, size = 1, position=position_dodge(width=0.5))+
    scale_color_manual(values = c("red", "bisque2", "skyblue"))+
  theme(legend.position = "bottom")+
  theme(axis.text.y = element_blank(), axis.title.y = element_blank())+
  xlab("Effect")


plot_tb_1<-ggplot(subset(calculate_effect_size_wide, Taxa != "uncultured"), aes(y = reorder(Taxa, Effect), x = Effect))+
  theme_bw()+
  geom_bar(stat = "identity", aes(fill = Effect))+
  scale_fill_viridis(discrete = F)+
  geom_bar(stat = "identity", aes(alpha = Significant), fill = "grey")+
  scale_alpha_discrete(range = c(1,0))+
  theme(legend.position = "bottom")+
  ylab("Genus") +
  theme(panel.grid.minor = element_line( size=0.5), panel.grid.major = element_blank()) 

ggarrange(plot_tb_1, plot_tb_2, ncol = 2, align = "h", legend = "bottom", widths = c(1.5,1))

```

-   TB estimates when year is a random effect

```{r}

TB_estimates<-list()
Condition_estimates<-list()


taxanames<-unique(top_genera_melt2$Genus)

for (i in 1:length(taxanames)){
  tryCatch({ #catch errors
    print(i)
  
    taxa1<-subset(top_genera_melt2, Genus == taxanames[i])
   # taxa1<-subset(top_genera_melt2, Genus == "Blautia")
    
    metadata<-taxa1
    #str(metadata)
    
    #scale variables
    metadata$Seq_depth_nospikein<-as.numeric(scale(metadata$Seq_depth_nospikein))
    metadata$TB_stage<-factor(metadata$TB_stage, levels = c("Unexposed", "Exposed","Diseased"))
    
    #str(metadata)
    metadata$Year<-factor(metadata$Year)
    
    # fit gamm    
    m_taxa <- mgcv::bam(Abundance~
        s(HoursAfterSunrise, bs = "cr")+
        s(month, bs = "cc")+
        s(AgeAtSamplingYears, bs = "cr")+
        s(Seq_depth_nospikein, bs = "cr")+
       # s(Year, bs = "cr", k = 5)+
        # s(TotalAbundance, bs = "cr")+
        TB_stage+
         Condition_weight+
        Storage+
        Seq_run+
        s(IndividID, bs = "re")+
        s(Year, bs = "re")+
         s(GroupAtSampling, bs = "re"),
      #correlation = corARMA(form = ~ 1|Year, p = 3),
      data=metadata,
      family = gaussian)
    
   print(summary(m_taxa))
   

       summary<-summary(m_taxa)
     
    ## TB estimates 
    
    TB_estimates_df<-data.frame(summary$p.table)[1:3,]
     TB_estimates_df$Level<-c("Unexposed", "Exposed", "Diseased")
     names(TB_estimates_df)[4]<-"P_val"
     
    TB_estimates_df$Estimate2<-ifelse(TB_estimates_df$Level == "Unexposed", 0, TB_estimates_df$Estimate)
         
     
      TB_estimates_df$lower<-NA
      TB_estimates_df$upper<-NA
    
      # confidence intervals exposed
      beta <- coef(m_taxa)
      Vb <- vcov(m_taxa, unconditional = TRUE)
     se <- sqrt(diag(Vb))
      j <- which(names(beta) == "(Intercept)")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[1,7]<-cis[1]
      TB_estimates_df[1,8]<-cis[2]
      
            # confidence intervals exposed
      j <- which(names(beta) == "TB_stageExposed")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[2,7]<-cis[1]
      TB_estimates_df[2,8]<-cis[2]
     
                 # confidence intervals symptomatic
      j <- which(names(beta) == "TB_stageDiseased")
    cis<- beta[j] + (c(-1,1) * (2 * se[j]))
     
      TB_estimates_df[3,7]<-cis[1]
      TB_estimates_df[3,8]<-cis[2]
      
      ## add coefficient
      
      TB_estimates_df$Coef <-as.numeric(coef(m_taxa)[1:3])
        TB_estimates_df$Taxa <-taxanames[i]
    TB_estimates[[i]]<-TB_estimates_df
      
      ## extract condition stats
      
      condition_estimates<- data.frame(summary$p.table)[4,]
       condition_estimates$Taxa<-taxanames[i]
       Condition_estimates[[i]]<-condition_estimates
      
  
    
  }, error=function(e){})
}

##########
##########
##########
##########

names(TB_estimates)<-taxanames

TB_estimates_df<-do.call(rbind, TB_estimates)

head(TB_estimates_df)

TB_estimates_df$Significant <-TB_estimates_df$P_val < 0.05

# edit significance (intercept will always be significant)

sig_taxa<-subset(TB_estimates_df, Significant == T & Level !="Unexposed")
sig_taxa<-unique(sig_taxa$Taxa)

TB_estimates_df$Significant<- TB_estimates_df$Taxa %in%  sig_taxa


TB_estimates_df

## edit confidence intervals

TB_estimates_df$Level<-factor(TB_estimates_df$Level, levels = c("Diseased", "Exposed", "Unexposed"))

TB_estimates_df$lower<- ifelse(TB_estimates_df$Level == "Unexposed", TB_estimates_df$lower-TB_estimates_df$Estimate, TB_estimates_df$lower)

TB_estimates_df$upper<- ifelse(TB_estimates_df$Level == "Unexposed", TB_estimates_df$upper-TB_estimates_df$Estimate, TB_estimates_df$upper)


# order taxa

TB_estimates_df$Taxa<-factor(TB_estimates_df$Taxa, levels = taxa_order)

### plot ###########
### plot ###########
### plot ###########
### plot ###########


# first add class column so can match colours with earlier figs

head(calculate_effect_size_wide)

taxonomy<-data.frame(tax_table(meerkat_genus))
tax<-taxonomy[,c("Genus", "Class")]
tax<-distinct(tax, Genus, .keep_all = T)
head(tax)

calculate_effect_size_wide$Class <- vlookup(calculate_effect_size_wide$Taxa, tax, lookup_column = "Genus", result_column = "Class")

head(calculate_effect_size_wide)
calculate_effect_size_wide$Class_plot<-ifelse(calculate_effect_size_wide$Class %in% top_class, calculate_effect_size_wide$Class, "Other")

calculate_effect_size_wide$Class_plot<-factor(calculate_effect_size_wide$Class_plot, levels = c("Bacteroidia","Fusobacteriia" ,  "Other","Gammaproteobacteria", "Clostridia",  
"Coriobacteriia","Actinobacteria",    "Bacilli") )

## 
calculate_effect_size_wide$Sig <-ifelse(calculate_effect_size_wide$Significant==T, "yes", "no")


plot_tb_1<-ggplot(subset(calculate_effect_size_wide, Taxa != "uncultured"), aes(y = reorder(Taxa, Effect), x = Effect))+
  theme_bw(base_size = 14)+
  geom_bar_pattern(stat = "identity", aes(fill = Class_plot, pattern = Sig),
                   color = "black", 
                   pattern_fill = "black",
                   pattern_angle = 45,
                   pattern_density = 0.1,
                   pattern_spacing = 0.025,
                   pattern_key_scale_factor = 0.6)+
  
   scale_pattern_manual(values = c(no = "stripe", yes = "none"))+
  scale_fill_manual(values = rev(pal[-5]))+

  ylab("") +
  theme(plot.margin=unit(c(1,0.2,0.2,0.2),"cm"))+
  xlab("Effect size")+
  labs(pattern = "Significant")+
  labs(fill = "Bacterial class")



#############
#############
#############
#############

plot_tb_2<-ggplot(subset(TB_estimates_df, Taxa!= "uncultured"), aes( x = Estimate2, y = Taxa, group = Level))+
  theme_bw(base_size = 14)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  geom_errorbarh(aes(col = Level, xmin = Estimate2- Std..Error, xmax = Estimate2+ Std..Error), height = 0, size = 1, position=position_dodge(width=0.5))+
  
    geom_errorbarh(aes(alpha = Significant, xmin = Estimate2- Std..Error, xmax = Estimate2+ Std..Error), height = 0, size = 1, col = "grey", position=position_dodge(width=0.5))+
  
  scale_alpha_discrete(range = c(1,0), guide = "none")+

   geom_point(aes(col = Level), size = 2, position=position_dodge(width=0.5))+
  
  scale_color_manual(values = c("red", "burlywood1", "skyblue"),  guide = guide_legend(reverse = TRUE))+
  xlim(c(-1.3,NA))+

  theme(axis.text.y = element_blank(), axis.title.y = element_blank())+
  xlab("Effect size")+
  labs(col = "TB stage")+
  theme(plot.margin=unit(c(1,0.2,0.2,0.2),"cm"))+
  # legend
  theme( legend.background = element_blank(),
        legend.box.background = element_rect(colour = "grey", fill = "white"))+ theme(legend.position=c(0.2,0.91), legend.direction="vertical", 
        legend.key.height = unit(0.05, 'cm'),
        legend.key.width = unit(0.05, 'cm'),
        legend.title=element_text(size=10))+
   guides(colour=guide_legend(title.position="top",                           title.hjust =0.5,reverse = TRUE))

```

# Body condition

```{r}
names(Condition_estimates)<-taxanames

condition_estimates_df<-do.call(rbind, Condition_estimates)

condition_estimates_df$Significant <-condition_estimates_df$Pr...t.. < 0.05


condition_estimates_df$Taxa<-factor(condition_estimates_df$Taxa, levels = taxa_order)

ggplot(subset(condition_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  geom_point( aes(col = Significant), size = 3)+
  theme_bw()+
  geom_errorbarh(aes(col = Significant, xmin = Estimate- Std..Error, xmax = Estimate+ Std..Error), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "red"))
  



plot_condition<-
ggplot(subset(condition_estimates_df, Taxa!= "uncultured"), aes( x = Estimate, y = Taxa))+
  theme_bw(base_size = 14)+
  geom_point( aes(col = Significant), size = 3)+
  geom_errorbarh(aes(col = Significant, xmin = Estimate- Std..Error, xmax = Estimate+ Std..Error), height = 0, size = 1)+
  geom_vline(xintercept = 0, linetype = "dashed")+
  scale_color_manual(values = c("grey", "red"))+
   theme(axis.text.y = element_blank(), axis.title.y = element_blank())+
  xlab("Effect size")+
  xlim(c(-0.008,NA))+
  #legend
   theme( legend.background = element_blank(),
        legend.box.background = element_rect(colour = "grey", fill = "white"))+
    theme(legend.position=c(0.8,0.92), legend.direction="vertical", 
        legend.key.height = unit(0.4, 'cm'),
        legend.key.width = unit(0.2, 'cm'),
        legend.title=element_text(size=12))+
   guides(colour=guide_legend(title.position="top", reverse = T,
                                     title.hjust =0.5))


# Merging legends
legend_1 <- get_legend(plot_tb_1)
#legend_2 <- get_legend(plot_)

#legends <- ggarrange(legend_1, legend_2, nrow=2)

# Combining plots
rm_legend <- function(p){p + theme(legend.position = "none")}

plots <- ggarrange(rm_legend(plot_tb_1), plot_tb_2, plot_condition, nrow = 1, align = "h", widths = c(1.8,1,1), labels = c("a)", "b)", "c)"))

# figure 3
ggarrange(plots, legend_1, widths = c(0.85, 0.15))

```

